---
title: 'Analysis 6: Directional Selection Analysis'
author: "Joe Gunn"
date: "2022-09-03"
output: html_document
---

# Project: Population genomic analysis of Smallmouth Bass and Neosho Bass in the Central Interior Highlands
We investigated the extent of genomic divergence, local directional selection, and admixture between the Smallmouth Bass (<i>Micropterus dolomieu</i>) and the Neosho Bass (<i>M. velox</i>) in the Central Interior Highlands (CIH) ecoregion of central north America. Specifically, we used ddRADseq data to assessed the phylogenomic relationship between and within species, characterizing inter- and intraspecific diversity and SNPs potentially under local directional selection at the population level. Additionally, we inferred the relative timing of admxiture in Neosho Bass streams where there is known introgressive hybridization with Smallmouth Bass to understand the influence of natural, historic geographic factors on mixing (stream capture or transient flooding) vs. anthropogenic factors (i.e., non-native introductions through stocking), which is known to have occurred widely in these economically valuable species. We ultimately hoped to provide novel insights into the diversity of endemic, ecologically important and popular sport fish in the CIH.

## Specific Aim: Directional selection analysis
In this analysis, we used the popgen.vcf data generated in Analysis 2 (SNP Filtering...) to scan for signatures of directional selection on SNP loci with outlier Fst (high outlier Fst: directional selection; low Fst: balancing selection). We used two software programs with different underlying statistical frameworks to detect outliers and then used any outliers commonly detected in both analyses as canditates for being under strong selection. Specifically, we used the software program BAYESCAN (based in Bayesian analysis) and the R package PCAdapt principal component analysis (based in multivariate principal component analysis). We then employed DAPC in R to map patterns of population differentiation at any shared outlier and neutral loci to detect populations that may be under differential selection pressures and to detect signatures of genetic drift, respectively.

## Phases of analysis:
### Phase 1: Data preparation
### Phase 2: 

## Libraries needed for analysis
```{r setup, echo = FALSE, include=FALSE}
library(readxl)
library(tidyverse)
library(cowplot)
library(devtools)
library(SNPRelate)
library(vcfR)
library(pophelper)
library(PopGenome)
library(pcadapt)
library(viridis)
```

### PHASE 1: DATA PREPARATION
In this phase, we are converting the fully filtered, thinned dataset excluding Spotted Bass and the Spotted Bass hybrid (`admixture_phylogenomics_analysis/data/processed_vcf/03_popgen_spb.vcf`) into input files compatible with the software programs BAYESCAN and the R package PCAdapt. For BAYESCAN, we are using the intermediate conversion software PGDSPIDER to generate input files, and for PCAdapt, we are again using the program PLINK.

For the outlier Fst analysis, we are conducing scans for outlier loci hiearchically at the black bass species level (between Spotted Bass, Smallmouth Bass, and Neosho Bass), among species within the Smallmouth Bass species complex (Smallmouth Bass and Neosho Bass), and among populations within black bass species. The goal with this structure is to reduce false positive detection due to population structure and to detect potentially selected SNPs at each hierarchical level.

With this analysis structure, we conducted scans with the following sets of data: 

1) All black bass samples, including Spotted Bass, Smallmouth Bass, and Neosho Bass (excluding the Spotted Bass hybrid)
2) All Smallmouth Bass and Neosho Bass samples, excluding Spotted Bass
3) All Neosho Bass populations only
4) All Smallmouth Bass populations only

All analyses conducted by calling commands in bash shell scripts were performed on the Lewis high-performance computing cluster at the University of Missouri <br>

<b>Programs needed</b>: <br><br>

PGDSPIDER v.2.1.1.5 (Lischer and Excoffier 2012) <br>

<b>Citation</b>:<br>

Lischer, H. E. L., E& Excoffier, L. (2012). PGDSpider: an automated data conversion tool fo connecting population genetics and genomics programs. Bionformatics, 28, 298-299. doi:/10.1093/bioinformatics/btr642


PLINK v.1.90p (high-contig build; Change et al. 2015) <br>

<b>Citation</b>:<br>

Chang, C. C., Chow, C. C., Tellier, L. C. A. M., Vattikuti, S., Purcell, S. M., & Lee, J. J. (2015). Second-generation PLINK: Rising to the challenge of larger and richer datasets. GigaScience, 4, 1â€“16. doi:10.1186/s13742-015-0047-8

### STEP 1: Prepare input data for BAYESCAN.
In this step, we are using the conversion software PGDSPIDER to prepare all input data for the software program BAYESCAN.

#### 1a: Generate PGDSPIDER accompanying SPID files
In this step, we are generating text scripts for an accompanying file needed for PGDSPIDER conversions called a "spid" file. This file includes specific parameters that must to be completed for each file conversion in order to generate the correct output.

##### 1a.1. Generate a general text script for PGDSPIDER SPID file
Here, we only provide the base script without any parameters filled in. In subsequent steps, we include the completed version for each file conversion with specific parameters. The base script is called "spid.spid".

###### 1a.1.1. Copy and paste the code below in a new shell script file:

##### Generate base SPID text script: `code/spid_files/spid.spid`. UNCOMMENT every other (all caps) line and fill in with the necessary parameters for each file conversion.
```{bash}
# VCF Parser questions
#PARSER_FORMAT=

# Do you want to include a file with population definitions?
#VCF_PARSER_POP_QUESTION=
# Only input following regions (refSeqName:start:end, multiple regions: whitespace separated):
#VCF_PARSER_REGION_QUESTION=
# What is the ploidy of the data?
#VCF_PARSER_PLOIDY_QUESTION=
# Only output following individuals (ind1, ind2, ind4, ...):
#VCF_PARSER_IND_QUESTION=
# Output genotypes as missing if the read depth of a position for the sample is below:
#VCF_PARSER_READ_QUESTION=
# Take most likely genotype if "PL" or "GL" is given in the genotype field?
#VCF_PARSER_PL_QUESTION=
# Do you want to exclude loci with only missing data?
#VCF_PARSER_EXC_MISSING_LOCI_QUESTION=
# Select population definition file:
#VCF_PARSER_POP_FILE_QUESTION=
# Only output SNPs with a phred-scaled quality of at least:
#VCF_PARSER_QUAL_QUESTION=
# Do you want to include non-polymorphic SNPs?
#VCF_PARSER_MONOMORPHIC_QUESTION=
# Output genotypes as missing if the phred-scale genotype quality is below:
#VCF_PARSER_GTQUAL_QUESTION=

# BayeScan Parser Questions
#WRITER_FORMAT=

# What type of data are we dealing with?
#GESTE_BAYE_SCAN_WRITER_DATA_TYPE_QUESTION=
```

##### 1a.2. Generate a PGDSPIDER SPID file for the black bass level genome scan
Here, we are filling in the parameters necessary to generate a spid file to convert input data to BAYESCAN format the Black Bass level genome scan, which includes all Spotted Bass, Smallmouth Bass, and Neosho Bass.

###### 1a.2.1. Copy and paste the code below in a new shell script file:

##### Generate base SPID text script: `code/spid_files/01_all.spid`. UNCOMMENT every other (all caps) line and fill in with the necessary parameters for each file conversion.
```{bash}
# VCF Parser questions
#PARSER_FORMAT=VCF

# Do you want to include a file with population definitions?
#VCF_PARSER_POP_QUESTION=true
# Only input following regions (refSeqName:start:end, multiple regions: whitespace separated):
#VCF_PARSER_REGION_QUESTION=
# What is the ploidy of the data?
#VCF_PARSER_PLOIDY_QUESTION=DIPLOID
# Only output following individuals (ind1, ind2, ind4, ...):
#VCF_PARSER_IND_QUESTION=
# Output genotypes as missing if the read depth of a position for the sample is below:
#VCF_PARSER_READ_QUESTION=
# Take most likely genotype if "PL" or "GL" is given in the genotype field?
#VCF_PARSER_PL_QUESTION=false
# Do you want to exclude loci with only missing data?
#VCF_PARSER_EXC_MISSING_LOCI_QUESTION=false
# Select population definition file:
#VCF_PARSER_POP_FILE_QUESTION=../../data/pop_files/01_all_pops.txt
# Only output SNPs with a phred-scaled quality of at least:
#VCF_PARSER_QUAL_QUESTION=
# Do you want to include non-polymorphic SNPs?
#VCF_PARSER_MONOMORPHIC_QUESTION=false
# Output genotypes as missing if the phred-scale genotype quality is below:
#VCF_PARSER_GTQUAL_QUESTION=

# BayeScan Parser Questions
#WRITER_FORMAT=GESTE_BAYE_SCAN

# What type of data are we dealing with?
#GESTE_BAYE_SCAN_WRITER_DATA_TYPE_QUESTION=SNP
```

##### 1a.3. Generate a PGDSPIDER SPID file for the Smallmouth Bass species complex genome scan
Here, we are filling in the parameters necessary to generate a spid file to convert input data to BAYESCAN format the species complex level genome scan, which includes all Smallmouth Bass and Neosho Bass.

###### 1a.3.1. Copy and paste the code below in a new shell script file:

##### Generate base SPID text script: `code/spid_files/02_species_complex.spid`. UNCOMMENT every other (all caps) line and fill in with the necessary parameters for each file conversion.
```{bash}
# VCF Parser questions
#PARSER_FORMAT=VCF

# Do you want to include a file with population definitions?
#VCF_PARSER_POP_QUESTION=true
# Only input following regions (refSeqName:start:end, multiple regions: whitespace separated):
#VCF_PARSER_REGION_QUESTION=
# What is the ploidy of the data?
#VCF_PARSER_PLOIDY_QUESTION=DIPLOID
# Only output following individuals (ind1, ind2, ind4, ...):
#VCF_PARSER_IND_QUESTION=
# Output genotypes as missing if the read depth of a position for the sample is below:
#VCF_PARSER_READ_QUESTION=
# Take most likely genotype if "PL" or "GL" is given in the genotype field?
#VCF_PARSER_PL_QUESTION=false
# Do you want to exclude loci with only missing data?
#VCF_PARSER_EXC_MISSING_LOCI_QUESTION=false
# Select population definition file:
#VCF_PARSER_POP_FILE_QUESTION=../../data/pop_files/02_species_complex_pops.txt
# Only output SNPs with a phred-scaled quality of at least:
#VCF_PARSER_QUAL_QUESTION=
# Do you want to include non-polymorphic SNPs?
#VCF_PARSER_MONOMORPHIC_QUESTION=false
# Output genotypes as missing if the phred-scale genotype quality is below:
#VCF_PARSER_GTQUAL_QUESTION=

# BayeScan Parser Questions
#WRITER_FORMAT=GESTE_BAYE_SCAN

# What type of data are we dealing with?
#GESTE_BAYE_SCAN_WRITER_DATA_TYPE_QUESTION=SNP
```

##### 1a.4. Generate a PGDSPIDER SPID file for the Neosho Bass genome scan
Here, we are filling in the parameters necessary to generate a spid file to convert input data to BAYESCAN format the Neosho Bass population level genome scan, which includes all Neosho Bass only

###### 1a.4.1. Copy and paste the code below in a new shell script file:

##### Generate base SPID text script: `code/spid_files/03_neosho.spid`. UNCOMMENT every other (all caps) line and fill in with the necessary parameters for each file conversion.
```{bash}
# VCF Parser questions
#PARSER_FORMAT=VCF

# Do you want to include a file with population definitions?
#VCF_PARSER_POP_QUESTION=true
# Only input following regions (refSeqName:start:end, multiple regions: whitespace separated):
#VCF_PARSER_REGION_QUESTION=
# What is the ploidy of the data?
#VCF_PARSER_PLOIDY_QUESTION=DIPLOID
# Only output following individuals (ind1, ind2, ind4, ...):
#VCF_PARSER_IND_QUESTION=
# Output genotypes as missing if the read depth of a position for the sample is below:
#VCF_PARSER_READ_QUESTION=
# Take most likely genotype if "PL" or "GL" is given in the genotype field?
#VCF_PARSER_PL_QUESTION=false
# Do you want to exclude loci with only missing data?
#VCF_PARSER_EXC_MISSING_LOCI_QUESTION=false
# Select population definition file:
#VCF_PARSER_POP_FILE_QUESTION=../../data/pop_files/03_neosho_pops.txt
# Only output SNPs with a phred-scaled quality of at least:
#VCF_PARSER_QUAL_QUESTION=
# Do you want to include non-polymorphic SNPs?
#VCF_PARSER_MONOMORPHIC_QUESTION=false
# Output genotypes as missing if the phred-scale genotype quality is below:
#VCF_PARSER_GTQUAL_QUESTION=

# BayeScan Parser Questions
#WRITER_FORMAT=GESTE_BAYE_SCAN

# What type of data are we dealing with?
#GESTE_BAYE_SCAN_WRITER_DATA_TYPE_QUESTION=SNP
```

##### 1a.5. Generate a PGDSPIDER SPID file for the Neosho Bass genome scan
Here, we are filling in the parameters necessary to generate a spid file to convert input data to BAYESCAN format the Smallmouth Bass population level genome scan, which includes all Smallmouth Bass only

###### 1a.5.1. Copy and paste the code below in a new shell script file:

##### Generate base SPID text script: `code/spid_files/04_smallmouth.spid`. UNCOMMENT every other (all caps) line and fill in with the necessary parameters for each file conversion.
```{bash}
# VCF Parser questions
#PARSER_FORMAT=VCF

# Do you want to include a file with population definitions?
#VCF_PARSER_POP_QUESTION=true
# Only input following regions (refSeqName:start:end, multiple regions: whitespace separated):
#VCF_PARSER_REGION_QUESTION=
# What is the ploidy of the data?
#VCF_PARSER_PLOIDY_QUESTION=DIPLOID
# Only output following individuals (ind1, ind2, ind4, ...):
#VCF_PARSER_IND_QUESTION=
# Output genotypes as missing if the read depth of a position for the sample is below:
#VCF_PARSER_READ_QUESTION=
# Take most likely genotype if "PL" or "GL" is given in the genotype field?
#VCF_PARSER_PL_QUESTION=false
# Do you want to exclude loci with only missing data?
#VCF_PARSER_EXC_MISSING_LOCI_QUESTION=false
# Select population definition file:
#VCF_PARSER_POP_FILE_QUESTION=../../data/pop_files/04_smallmouth_pops.txt
# Only output SNPs with a phred-scaled quality of at least:
#VCF_PARSER_QUAL_QUESTION=
# Do you want to include non-polymorphic SNPs?
#VCF_PARSER_MONOMORPHIC_QUESTION=false
# Output genotypes as missing if the phred-scale genotype quality is below:
#VCF_PARSER_GTQUAL_QUESTION=

# BayeScan Parser Questions
#WRITER_FORMAT=GESTE_BAYE_SCAN

# What type of data are we dealing with?
#GESTE_BAYE_SCAN_WRITER_DATA_TYPE_QUESTION=SNP
```

#### 1b: Generate population designation files (read by SPID files)
In this step, we are generating text files listing all samples in at each genome scan level and their corresponding population designations as inferred in Analysis 4 (population inference). Our inferred populations were named according to their general geographic locations (see Phase 4 of population_analysis, "MIDARK", "ELK", "BAYOU", etc.), but PGDSPIDER expects input populations in a specific format (i.e., "pop_1", "pop_2", etc.). We generate population files according to this naming convention

##### 1b.1. Read in sample metadata and generate lists of populations for the black bass species level.

###### 1b.1.1 Generate population file for all black bass samples (Spotted Bass, Smallmouth Bass, and Neosho Bass):

##### Generate population file: `data/pop_files/01_all_pops.txt`
```{r}
# Load in full sample metadata file with population designations (inferred in Analysis 4, population_analysis)
load("../population_analysis/data/metadata/metadata_populations_spb.Rda")

metadata_populations_spb <- metadata_populations_spb %>%
  mutate(population = as.character(population))

# Rename population designations in SPID format ("pop_1", "pop_2", etc.)
metadata_populations_spb$population[metadata_populations_spb$population == "MIDARK"] <- "pop_1"
metadata_populations_spb$population[metadata_populations_spb$population == "MISS"] <- "pop_2"
metadata_populations_spb$population[metadata_populations_spb$population == "SKIA"] <- "pop_3"
metadata_populations_spb$population[metadata_populations_spb$population == "LMULB"] <- "pop_4"
metadata_populations_spb$population[metadata_populations_spb$population == "WHITE"] <- "pop_5"
metadata_populations_spb$population[metadata_populations_spb$population == "ELK"] <- "pop_6"
metadata_populations_spb$population[metadata_populations_spb$population == "ILLI"] <- "pop_7"
metadata_populations_spb$population[metadata_populations_spb$population == "UPPARK"] <- "pop_8"
metadata_populations_spb$population[metadata_populations_spb$population == "BAYOU"] <- "pop_9"
metadata_populations_spb$population[metadata_populations_spb$population == "SPB"] <- "pop_10"

# Generate text file
write_tsv(metadata_populations_spb, 
          file = "data/pop_files/01_all_pops.txt", 
          col_names = FALSE)
```

##### 1b.2. Read in sample metadata and generate lists of populations for the species complex level.

###### 1b.2.1 Generate population file for species complex samples (Smallmouth Bass and Neosho Bass):

##### Generate population file: `data/pop_files/02_species_complex_pops.txt`
```{r}
# Load in full sample metadata file with population designations (inferred in Analysis 4, population_analysis)
load("../population_analysis/data/metadata/metadata_populations_spb.Rda")

metadata_populations_complex <- metadata_populations_spb %>%
  filter(population != "SPB")

metadata_populations_complex <- metadata_populations_complex %>%
  mutate(population = as.character(population))

# Rename population designations in SPID format ("pop_1", "pop_2", etc.)
metadata_populations_complex$population[metadata_populations_complex$population == "MIDARK"] <- "pop_1"
metadata_populations_complex$population[metadata_populations_complex$population == "MISS"] <- "pop_2"
metadata_populations_complex$population[metadata_populations_complex$population == "SKIA"] <- "pop_3"
metadata_populations_complex$population[metadata_populations_complex$population == "LMULB"] <- "pop_4"
metadata_populations_complex$population[metadata_populations_complex$population == "WHITE"] <- "pop_5"
metadata_populations_complex$population[metadata_populations_complex$population == "ELK"] <- "pop_6"
metadata_populations_complex$population[metadata_populations_complex$population == "ILLI"] <- "pop_7"
metadata_populations_complex$population[metadata_populations_complex$population == "UPPARK"] <- "pop_8"
metadata_populations_complex$population[metadata_populations_complex$population == "BAYOU"] <- "pop_9"

# Generate text file
write_tsv(metadata_populations_complex, 
          file = "data/pop_files/02_species_complex_pops.txt", 
          col_names = FALSE)
```

##### 1b.3. Read in sample metadata and generate lists of populations for the Neosho Bass level.

###### 1b.3.1 Generate population file for Neosho Bass samples (Neosho Bass only):

##### Generate population file: `data/pop_files/03_neosho_pops.txt`
```{r}
# Load in full sample metadata file with population designations (inferred in Analysis 4, population_analysis)
load("../population_analysis/data/metadata/metadata_populations_spb.Rda")

metadata_populations_neosho <- metadata_populations_spb %>%
  filter(population != "SPB") %>%
  filter(population != "MISS") %>%
  filter(population != "SKIA") %>%
  filter(population != "WHITE")

metadata_populations_neosho <- metadata_populations_neosho %>%
  mutate(population = as.character(population))

# Rename population designations in SPID format ("pop_1", "pop_2", etc.)
metadata_populations_neosho$population[metadata_populations_neosho$population == "MIDARK"] <- "pop_1"
metadata_populations_neosho$population[metadata_populations_neosho$population == "LMULB"] <- "pop_2"
metadata_populations_neosho$population[metadata_populations_neosho$population == "ELK"] <- "pop_3"
metadata_populations_neosho$population[metadata_populations_neosho$population == "ILLI"] <- "pop_4"
metadata_populations_neosho$population[metadata_populations_neosho$population == "UPPARK"] <- "pop_5"
metadata_populations_neosho$population[metadata_populations_neosho$population == "BAYOU"] <- "pop_6"

# Generate text file
write_tsv(metadata_populations_neosho, 
          file = "data/pop_files/03_neosho_pops.txt", 
          col_names = FALSE)
```

##### 1b.4. Read in sample metadata and generate lists of populations for the Smallmouth Bass level.

###### 1b.4.1 Generate population file for Smallmouth Bass samples (Smallmouth Bass only):

##### Generate population file: `data/pop_files/04_neosho_pops.txt`
```{r}
# Load in full sample metadata file with population designations (inferred in Analysis 4, population_analysis)
load("../population_analysis/data/metadata/metadata_populations_spb.Rda")

metadata_populations_smallmouth <- metadata_populations_spb %>%
  filter(population != "SPB") %>%
  filter(population != "MIDARK") %>%
  filter(population != "ELK") %>%
  filter(population != "ILLI") %>%
  filter(population != "LMULB") %>%
  filter(population != "BAYOU") %>%
  filter(population != "UPPARK")

metadata_populations_smallmouth <- metadata_populations_smallmouth %>%
  mutate(population = as.character(population))

# Rename population designations in SPID format ("pop_1", "pop_2", etc.)
metadata_populations_smallmouth$population[metadata_populations_smallmouth$population == "WHITE"] <- "pop_1"
metadata_populations_smallmouth$population[metadata_populations_smallmouth$population == "MISS"] <- "pop_2"
metadata_populations_smallmouth$population[metadata_populations_smallmouth$population == "SKIA"] <- "pop_3"

# Generate text file
write_tsv(metadata_populations_smallmouth, 
          file = "data/pop_files/04_smallmouth_pops.txt", 
          col_names = FALSE)
```

#### 1c: Generate a general bash shell script header for all PGDSPIDER conversion commands
In this step, we are generating a universal bash shell script to run all file type conversions in PGDSPIDER. Here, we only generate the annotated header for the shell script, and in subsequent steps, we include and provide details on each file conversion and the associated lines of code in Rmd chunks. The shell script for running all PGDSPIDER code is called "pgdspider.sh".

##### 1c.1: Copy and paste the code below in a new shell script file:

##### Generate PGDSPIDER bash shell script: `code/shell_scripts/pgdspider.sh`
```{bash}
#! /bin/bash

#SBATCH -p Lewis  # use the Lewis partition
#SBATCH -J pgd_coversion  # give the job a custom name
#SBATCH -o pgd_results-%j.out  # give the job output a custom name
#SBATCH -t 0-02:00  # two hour time limit
#SBATCH --mem-per-cpu=8G

#SBATCH -N 1  # number of nodes
#SBATCH -n 1  # number of cores (AKA tasks)

# UNCOMMENT STARTING HERE

## Commands here run only on the first core
# module load rss/rss-2020
# module load pgdspider/pgdspider-2.1.1.5
```

#### 1d: Convert VCF files to BAYESCAN format.
In this step, we are using the general bash script for PGDSPIDER in addition to the four unique spid files for each analysis level to generate input files for BAYESCAN.

##### 1d.1. Copy and paste the code below in the shell file generated in Step 1c.1 above, which will generate a BAYESCAN file for the black bass species level dataset:

##### Command line code for generating BAYESCAN file for all samples. UNCOMMENT this code in the shell script:
```{bash}
## conversion 01: all black bass samples
#java -Xmx8g -Xms8g -jar ${PGDSPIDER_ROOT}/PGDSpider2-cli.jar -inputfile ../../data/processed_vcf/01_popgen_spb_hybrid.vcf -inputformat VCF -outputfile ../../data/processed_bayescan/01_all -outputformat GESTE_BAYE_SCAN -spid ../spid_files/01_all.spid
```

Run: `sbatch pgdspider.sh`

This code generates three files here: `data/processed_bayescan/`; these files include:

<b>Files</b>: <br><br>
1) popgen.ind
2) popgen.snp
3) popgen.eigenstratgeno

