---
title: "Analysis 3: Admixture and Phylogenomics"
author: "Joe Gunn"
date: "10/30/2019"
output: html_document
---

# Project: Population genomic analysis of Smallmouth Bass and Neosho Bass in the Central Interior Highlands
We investigated the extent of genomic divergence, local directional selection, and admixture between the Smallmouth Bass (<i>Micropterus dolomieu</i>) and the Neosho Bass (<i>M. velox</i>) in the Central Interior Highlands (CIH) ecoregion of central north America. Specifically, we used ddRADseq data to assessed the phylogenomic relationship between and within species, characterizing inter- and intraspecific diversity and SNPs potentially under local directional selection at the population level. Additionally, we inferred the relative timing of admxiture in Neosho Bass streams where there is known introgressive hybridization with Smallmouth Bass to understand the influence of natural, historic geographic factors on mixing (stream capture or transient flooding) vs. anthropogenic factors (i.e., non-native introductions through stocking), which is known to have occurred widely in these economically valuable species. We ultimately hoped to provide novel insights into the diversity of endemic, ecologically important and popular sport fish in the CIH.

## Specific Aim: Admixture and phylogenomics analysis 
In this analysis, we used the popgen.vcf data generated in Analysis 2 (SNP Filtering...) to assess population genomic structure of Spotted Bass, Smallmouth Bass, and Neosho Bass in the CIH. Specifically, we conducted an initial screen of hybridization and gene flow by running maximum likelihood clustering on the full, filtered dataset and identified any individuals of interspecific origin between Spotted Bass and all other Interior Highlands fish (Smallmouth Bass and Neosho Bass) and individuals of interspecific origin between Smallmouth Bass and Neosho Bass. After removing hybrids, we conducted a separate admixture analysis and complementary phylogenomic analysis on the "pure" genomic samples to estimate genomic divergence between species.

## Phases of analysis:
### Phase 1: Data preparation
### Phase 2: Admixture analysis
### Phase 3: Phylogenomic analysis

## Libraries needed for analysis
```{r setup, echo = FALSE, include=FALSE}
library(readxl)
library(tidyverse)
library(cowplot)
library(devtools)
library(pophelper)
library(gitcreds)
```

## PHASE 1: Metadata preparation
In this phase of the analysis, we reading in a metadata file that was prepared in excel prior to analysis (`metadata.xlsx`), which gives the sample ID, species designation (by native range or morphology in the case of Spotted Bass), and stream from which it was capture.

All analyses conducted by calling commands in bash shell scripts were performed on the Lewis high-performance computing cluster at the University of Missouri <br>

<b>Programs needed</b>: <br><br>

PLINK v.1.90p (high-contig build; Change et al. 2015) <br>

<b>Citation</b>:<br>

Chang, C. C., Chow, C. C., Tellier, L. C. A. M., Vattikuti, S., Purcell, S. M., & Lee, J. J. (2015). Second-generation PLINK: Rising to the challenge of larger and richer datasets. GigaScience, 4, 1–16. doi:10.1186/s13742-015-0047-8

### STEP 1: Prepare raw metadata for use in downstream analyses and summarize sample size data across streams.
In this step, we are reading in, cleaning, and filtering the raw metadata to prepare for downstream analyses.

#### 1a: Read in and clean metadata; run the Rmd chunk below:

##### Read in and clean metadata
```{r METADATA}
## Read in metadata
metadata <- read_excel("data/metadata/metadata.xlsx")

## Convert characters to factors
metadata <- metadata %>%
  mutate(sample_id = factor(sample_id),
         species = factor(species),
         stream = factor(stream))

## Omit samples removed from filtering steps in Analysis 2
metadata <- metadata %>%
  filter(sample_id != "GRSPB34") %>%
  filter(sample_id != "BFORK32") %>%
  filter(sample_id != "ER05")
```

#### 1b: Summarize sample sizes by species and stream; run the Rmd chunk below:

#### Summarize metadata
```{r}
## Get sample sizes by species
species_sample_sizes <- metadata %>%
  group_by(species) %>%
  count()

## Get sample sizes by stream
species_sample_sizes <- metadata %>%
  group_by(stream) %>%
  count()
```

These tables were used as the basis for maintext Table 1 in the manuscript (in the manuscript version of this table, we also include in the total sample sizes the samples filtered due to low genotype call rates).

### ----------------------- END OF PHASE 1: METADATA PREPARATION  ----------------------- ###

## PHASE 2: ADMIXTURE ANALYSIS
In this phase, we are reading in the previously filtered VCF data (`data/processed_vcf/popgen.vcf`), converting the data to input files necessary for the program ADMIXTURE, and running the following analyses: 1) a screen of the full dataset for interspecific hybrids between Spotted Bass and either Smallmouth Bass or Neosho Bass and a for interspecific hybrids between Smallmouth Bass and Neosho Bass; and 2) population genomic structure analysis of the genetically "pure" samples only, which will be combined with subsequent phylogenomic analysis of pure samples. <br>

All analyses conducted by calling commands in bash shell scripts were performed on the Lewis high-performance computing cluster at the University of Missouri <br>

<b>Programs needed</b>: <br><br>

ADMIXTURE v.1.3 (Alexander et al. 2009) <br>

Alexander, D. H., Novembre, J., & Lange, K. (2009). Fast model-based estimation of ancestry in unrelated individuals. Genome Research, 19, 1655–1664. doi:10.1101/gr.094052.109

### STEP 1: Generate a general bash shell script header for all PLINK commands
In this step, we are generating a universal bash shell script to run all file type conversions in PLINK. Here, we only generate the annotated header for the shell script, and in subsequent steps, we include and provide details on each file conversion and the associated lines of code in Rmd chunks. The shell script for running all PLINK code is called "plink.sh".

##### 1a: Generate the header for a general bash shell script to run PLINK:

##### Generate plink bash shell script: `code/shell_scripts/plink.sh`
```{bash}
#! /bin/bash

#SBATCH -p Lewis  # use the Lewis partition
#SBATCH -J plink_conversion # give the job a custom name
#SBATCH -o get_plink_bed-%j.out  # give the job output a custom name
#SBATCH -t 0-01:00  # two hour time limit

#SBATCH -N 1  # number of nodes
#SBATCH -n 1  # number of cores (AKA tasks)

# Commands here run only on the first core

# Modules to load

## module load plink/plink-high-contig-1.90p (UNCOMMENT THIS LINE TO RUN THE CODE)
```

### STEP 2: Screen for hybridization population genomic admixture analysis on the FULL dataset
In this step, we are running a population genomic analysis of the full, filtered VCF dataset to run a screen for interspecific hybridization, both between Spotted Bass and all other samples (Smallmouth Bass and Neosho Bass), as well as between Smallmouth Bass and Neosho Bass.

#### 2a: Convert VCF data into .bed format for admixture analysis in the software program ADMIXURE.

##### 2a.1. Copy and paste the code below in the shell file generated in Step 1a above, which will generate a .bed, along with .bim, .fam, .log, and .nosex files:

<b>We use the following flags in the code</b>: <br>
    --allow-extra-chr: this flag is necessary due to the data being in the format of contigs rather than chromosomes
    --vcf-half-call m: this flag treats any half genotype calls (e.g., '0/.' or './1') as missing data
    --make-bed: this specifies output format of .bed

##### Command line code for converting VCF to BED. UNCOMMENT this code in the shell script and run: `sbatch plink.sh`
```{bash}
#plink --vcf ../../data/processed_vcf/popgen.vcf --allow-extra-chr --vcf-half-call m --make-bed --out ../../processed_bed/popgen_all
```

This code generates the file: `data/processed_bed/all_samples/01_all.bed`, along with accompanying file types. This file should be used as input in analysis in ADMIXTURE for the full dataset.

### STEP 3: Screen full, filtered dataset for interspecific hybrids by conducting population genomic analysis in ADMIXTURE.

#### 3a: Generate a batch list file for running ADMIXTURE in parallel.
In this step, we are creating a batch list of command line code to run ADMIXTURE analyses in parallel. The batch list  contains a separate line of code to run a single replicate at an a priori determined number of populations (<i>K</i>). We are running this analysis with a random seed (--seed=2324), 10-fold cross validation (--cv=10), for <i>K</i> = 1-20 equalling the number of streams in our data set plus Spotted Bass. <br>

##### Generate batch commands: `code/batch_cmd_lists/01_all_batch_cmd_list.txt`
```{r}
nk <- data.frame(c(1:20))

cat("", file="code/batch_cmd_lists/01_all_batch_cmd_list.txt")

# Run loop to create file for storing commands
for(ii in 1:nrow(nk)) {
    
  admixture_call <- paste("admixture --seed=2324 --cv=10")

  input <- paste("../../data/processed_bed/all_samples/01_all.bed")
  
  cat(paste(admixture_call, input, nk[ii,]), 
      "\n", 
      file=paste("code/batch_cmd_lists/01_all_batch_cmd_list.txt"), # name file to store list
      append=TRUE)
  }

```

##### 3b: Generate a shell script to run ADMXITURE:
In this step, we are generating a universal bash shell script to run all population genomic analyses in ADMIXTURE. The shell script for running all ADMIXTURE code is called "admixture.sh" and is adapted in the steps below to run the specified input files at each step. 

##### Generate plink bash shell script: `code/shell_scripts/admixture.sh`
```{bash}
#! /bin/bash

#SBATCH --account=biosci
#SBATCH -p Lewis  # use the Lewis partition
#SBATCH -J admix # give the job a custom name
#SBATCH -o admix_results-%j.out  # give the job output a custom name
#SBATCH -t 0-03:00  # two hour time limit

#SBATCH -N 1  # number of nodes
#SBATCH -n 7  # number of cores (AKA tasks)

# Commands here run only on the first core

## UNCOMMENT LINES STARTING HERE 

# echo "### Starting at: $(date) ###"


## load packages (LEAVE THIS COMMENTED)
# module load rss/rss-2020
# module load admixture_linux/admixture_linux-1.3.

# COMMANDA=`head -n ${SLURM_ARRAY_TASK_ID} ../batch_cmd_lists/01_all_batch_cmd_list.txt | tail -n 1`
# eval $COMMANDA

# echo "### Ending at: $(date) ###"
```



##Only samples that were pure by original admixture analysis (this analysis is conducted below, but I create the metadata set here for consistency in the R workflow)
genomics_metadata_pure <- genomics_metadata %>%
   filter(adm_status == "pure")

#Filter data file for only Smallmouth Bass and INCLUDING the SMB X SPB (Spotted Bass) hybrid (BFC10)
genomics_metadata_hybrid_smb <- genomics_metadata %>%
   filter(species != "Spotted Bass")

genomics_metadata_hybrid_neosho <- genomics_metadata_hybrid_smb %>%
   filter(species != "Northern")

genomics_metadata_hybrid_northern <- genomics_metadata_hybrid_smb %>%
   filter(species != "Neosho")

#Filter data file for only Smallmouth Bass and EXCLUDING the SMB X SPB (Spotted Bass) hybrid (BFC10)
genomics_metadata_nohybrid_all <- genomics_metadata %>%
   filter(sample_id != "BFC10")

genomics_metadata_nohybrid_smb <- genomics_metadata %>% 
   filter(species != "Spotted Bass") %>% 
   filter(sample_id != "BFC10") #EXCLUDE THE HYBRID BFC10

#Filter data file for only Neosho Smallmouth Bass
genomics_metadata_nohybrid_neosho <- genomics_metadata_nohybrid_smb %>% 
   filter(species != "Northern")

genomics_metadata_nohybrid_northern <- genomics_metadata_nohybrid_smb %>%
   filter(species != "Neosho")

#Filter PURE samples data file for only Smallmouth Bass (Northern and Neosho) that are pure
genomics_metadata_pure_smb <- genomics_metadata_pure %>%
   filter(species != "Spotted Bass")

#Filter full dataset for just admixed individuals
genomics_metadata_admixed <- genomics_metadata %>%
   filter(adm_status == "admixed") %>%
   filter(sample_id != "BFC10")

```

# Hybrid Screen Analysis
### All visualization done using the library "pophelper"

## Screen for Species-level hybridization between Smallmouth Bass and Spotted Bass as well as admixture between Northern Smallmouth Bass and Neosho Smallmouth Bass
```{r}
#Read-in ADMIXTURE q-value results for all samples at best K value (best K = 4)
admix_spb_screen_K4 <- readQ("../../raw_data/admixture_data/admixture_q_data/all_samples/pass2_all.4.Q") # K=4 was the best supported value of K at CV = 0.253
admix_spb_screen_K3 <- readQ("../../raw_data/admixture_data/admixture_q_data/all_samples/pass2_all.3.Q")

#Plot ADMIXTURE results for K = 4 for all samples (92) to screen for spotted bass and smallmouth bass hybrids. Any hybrids will be removed from downstream analysis

##In order of POPULATION IDs (see pop_num column of metadata)
plotQ(admix_spb_screen_K4,
      showindlab = T, 
      useindlab=T,
      indlabsize = 8,
      indlabcol = "black",
      grplab = genomics_metadata,
      showgrplab = F,
      ordergrp = T,
      selgrp = "sample_id",
      subsetgrp =  c("SPRMO49","GRSPB52","GRSPB51","GLVR11","STOCK03","STOCK04","STOCK05","STOCK06","BP01","BP02","BP07","BP10","BP17","TBLR01","TBLR02","TBLR03","TBLR04","GRSPB35","GRSPB36","GRSPB37","GRSPB39","GRSPB41","GRSPB50","SKIA03","SKIA04","SKIA05","SKIA06","SKIA09","BC01","BC07","BC08","BC09","BC12","BC14","SC06","SC37","MI419","MI420","MI421","MI422","MI423","MI425","ER17","ER20","ER30","ER33","ER35","ER44","HC38","HC40","HC41","HC42","HC43","HC44","SPVW11","SPVW02","SPVW05","SPVW07","SPVW12","SPVW14","NOIS07","NOIS08","NOIS12","NOIS18","GLVR4","SPRMO11","SPRMO19","BFORK02","BFORK23","BFORK24","BFORK30","BFORK49","GRSPB02","GRSPB03","GRSPB69","CANEY15","CANEY16","BFC02","BFC06","BFC10","BFC49","AR16","AR18","AR19","AR21","AR29","AR30","AT02","AT05","AT08","AT09","AT12"),
      clustercol = c("forestgreen","goldenrod3","deepskyblue","deeppink2"),
      showlegend = F,
      legendmargin = c(2,2,2,2),
      showsp = F, 
      showdiv = T, 
      divsize = 0.5, 
      divcol = "black",  
      divtype = 1,
      legendkeysize = 10, 
      legendtextsize = 15, 
      linesize = 0.5, 
      pointsize = 4, 
      barbordercolour = "black",
      barbordersize = 0.2, 
      outputfilename = "../../visualization/admixture_figures/admixture_plots/SPB_X_SMB_screen_K4",
      exportpath = getwd(),
      imgtype = "pdf",
      imgoutput = "sep",
      height = 5, width = 30)

##In order of POPULATION IDs (see pop_num column of metadata)
plotQ(admix_spb_screen_K3,
      showindlab = T, 
      useindlab=T,
      indlabsize = 8,
      indlabcol = "black",
      grplab = genomics_metadata,
      showgrplab = F,
      ordergrp = T,
      selgrp = "sample_id",
      subsetgrp =  c("SPRMO49","GRSPB52","GRSPB51","GLVR11","STOCK03","STOCK04","STOCK05","STOCK06","BP01","BP02","BP07","BP10","BP17","TBLR01","TBLR02","TBLR03","TBLR04","GRSPB35","GRSPB36","GRSPB37","GRSPB39","GRSPB41","GRSPB50","SKIA03","SKIA04","SKIA05","SKIA06","SKIA09","BC01","BC07","BC08","BC09","BC12","BC14","SC06","SC37","MI419","MI420","MI421","MI422","MI423","MI425","ER17","ER20","ER30","ER33","ER35","ER44","HC38","HC40","HC41","HC42","HC43","HC44","SPVW11","SPVW02","SPVW05","SPVW07","SPVW12","SPVW14","NOIS07","NOIS08","NOIS12","NOIS18","GLVR4","SPRMO11","SPRMO19","BFORK02","BFORK23","BFORK24","BFORK30","BFORK49","GRSPB02","GRSPB03","GRSPB69","CANEY15","CANEY16","BFC02","BFC06","BFC10","BFC49","AR16","AR18","AR19","AR21","AR29","AR30","AT02","AT05","AT08","AT09","AT12"),
      clustercol = c("goldenrod3","deepskyblue","deeppink2"),
      showlegend = F,
      legendmargin = c(2,2,2,2),
      showsp = F, 
      showdiv = T, 
      divsize = 0.5, 
      divcol = "black",  
      divtype = 1,
      legendkeysize = 10, 
      legendtextsize = 15, 
      linesize = 0.5, 
      pointsize = 4, 
      barbordercolour = "black",
      barbordersize = 0.2, 
      outputfilename = "../../visualization/admixture_figures/admixture_plots/SPB_X_SMB_screen_K3",
      imgtype = "pdf",
      imgoutput = "sep",
      height = 5, width = 30)
```

# Determine Major Lineages (Run only pure samples, no hybridization or admixture)

## All black bass pure samples (SPB, SMB)
```{r}
##Read in Admixture results (Q files)
all_pure_3 <- readQ("../../admixture_data/admixture_q_data/pure_samples/all_pure.3.Q")
all_pure_4 <- readQ("../../admixture_data/admixture_q_data/pure_samples/all_pure.4.Q")
all_pure_5 <- readQ("../../admixture_data/admixture_q_data/pure_samples/all_pure.5.Q")

#Make an admixture plot with 3 clusters, but in the order of snphylo results.
plotQ(all_pure_3,
      showindlab = T, 
      useindlab=T,
      indlabsize = 8,
      indlabcol = "black",
      grplab = genomics_metadata_pure,
      showgrplab = F,
      ordergrp = T,
      selgrp = "sample_id",
      subsetgrp =  c("GRSPB51","GRSPB52","SPRMO49","GLVR11","TBLR02","TBLR04","TBLR01","TBLR03","GRSPB37","GRSPB41","GRSPB39","GRSPB50","GRSPB35","GRSPB36","STOCK04","STOCK03","STOCK06","STOCK05","BP01","BP02","BP07","BP10","BP17","SKIA04","SKIA09","SKIA03","SKIA06","SKIA05","SC37","SC06","HC43","HC40","HC38","HC41","HC44","HC42","BFORK23","CANEY16","GRSPB69","BFORK02","GRSPB02","GRSPB03","BFORK49","NOIS12","BFC06","BFC49","BFC02","AR18","AR19","AR21","AR16"),
      clustercol = c("deeppink2","goldenrod3","deepskyblue"),
      showlegend = F,
      legendmargin = c(2,2,2,2),
      showsp = F, 
      showdiv = T, 
      divsize = 0.5, 
      divcol = "black",  
      divtype = 1,
      legendkeysize = 10, 
      legendtextsize = 15, 
      linesize = 0.5, 
      pointsize = 4, 
      barbordercolour = "black",
      barbordersize = 0.2, 
      outputfilename = "../../visualization/admixture_figures/admixture_plots/pure_lineages_3_snphylo_order",
      imgtype = "pdf",
      imgoutput = "sep",
      height = 5, width = 30)

#Make an admixture plot with 4 clusters, but in the order of snphylo results.
plotQ(all_pure_4,
      showindlab = T, 
      useindlab=T,
      indlabsize = 8,
      indlabcol = "black",
      grplab = genomics_metadata_pure,
      showgrplab = F,
      ordergrp = T,
      selgrp = "sample_id",
      subsetgrp =  c("GRSPB51","GRSPB52","SPRMO49","GLVR11","TBLR02","TBLR04","TBLR01","TBLR03","GRSPB37","GRSPB41","GRSPB39","GRSPB50","GRSPB35","GRSPB36","STOCK04","STOCK03","STOCK06","STOCK05","BP01","BP02","BP07","BP10","BP17","SKIA04","SKIA09","SKIA03","SKIA06","SKIA05","SC37","SC06","HC43","HC40","HC38","HC41","HC44","HC42","BFORK23","CANEY16","GRSPB69","BFORK02","GRSPB02","GRSPB03","BFORK49","NOIS12","BFC06","BFC49","BFC02","AR18","AR19","AR21","AR16"),
      clustercol = c("goldenrod3","deeppink2","deepskyblue","navyblue"),
      showlegend = F,
      legendmargin = c(2,2,2,2),
      showsp = F, 
      showdiv = T, 
      divsize = 0.5, 
      divcol = "black",  
      divtype = 1,
      legendkeysize = 10, 
      legendtextsize = 15, 
      linesize = 0.5, 
      pointsize = 4, 
      barbordercolour = "black",
      barbordersize = 0.2, 
      outputfilename = "../../visualization/admixture_figures/admixture_plots/pure_lineages_4_snphylo_order",
      imgtype = "pdf",
      imgoutput = "sep",
      height = 5, width = 30)

#Make an admixture plot with 5 clusters, but in the order of snphylo results.
plotQ(all_pure_5,
      showindlab = T, 
      useindlab=T,
      indlabsize = 8,
      indlabcol = "black",
      grplab = genomics_metadata_pure,
      showgrplab = F,
      ordergrp = T,
      selgrp = "sample_id",
      subsetgrp =  c("GRSPB51","GRSPB52","SPRMO49","GLVR11","TBLR02","TBLR04","TBLR01","TBLR03","GRSPB37","GRSPB41","GRSPB39","GRSPB50","GRSPB35","GRSPB36","STOCK04","STOCK03","STOCK06","STOCK05","BP01","BP02","BP07","BP10","BP17","SKIA04","SKIA09","SKIA03","SKIA06","SKIA05","SC37","SC06","HC43","HC40","HC38","HC41","HC44","HC42","BFORK23","CANEY16","GRSPB69","BFORK02","GRSPB02","GRSPB03","BFORK49","NOIS12","BFC06","BFC49","BFC02","AR18","AR19","AR21","AR16"),
      clustercol = c("deeppink2","goldenrod3","forestgreen","deepskyblue","navyblue"),
      showlegend = F,
      legendmargin = c(2,2,2,2),
      showsp = F, 
      showdiv = T, 
      divsize = 0.5, 
      divcol = "black",  
      divtype = 1,
      legendkeysize = 10, 
      legendtextsize = 15, 
      linesize = 0.5, 
      pointsize = 4, 
      barbordercolour = "black",
      barbordersize = 0.2, 
      outputfilename = "../../visualization/admixture_figures/admixture_plots/pure_lineages_5_snphylo_order",
      imgtype = "pdf",
      imgoutput = "sep",
      height = 5, width = 30)
```

## Determine if there is any population structure among the admixed individuals (within Neosho Smallmouth Bass)

## Only Northern and Neosho Smallmouth Bass samples (SMB)
```{r}
##Read in Admixture Results
admixed_2 <- readQ("../admixture_data/admixture_q_data/admixed_samples/admixed_indivs.2.Q")
admixed_3 <- readQ("../admixture_data/admixture_q_data/admixed_samples/admixed_indivs.3.Q") 
admixed_4 <- readQ("../admixture_data/admixture_q_data/admixed_samples/admixed_indivs.4.Q")

plotQ(admixed_2,
      showindlab = T, 
      useindlab=T,
      indlabsize = 8,
      indlabcol = "black",
      grplab = genomics_metadata_admixed,
      showgrplab = F,
      ordergrp = T,
      selgrp = "sample_id",
      subsetgrp =  c("NOIS18","NOIS08","NOIS07","BFORK30","BFORK24","SPRMO11","GLVR4","SPRMO19","CANEY15","AT12","AT05","AT08","AT09","AT02","AR29","AR30","ER33","ER35","ER20","ER30","ER44","ER17","BC01","BC08","BC09","BC12","SPVW14","SPVW07","MI422","MI421","SPVW02","SPVW05","SPVW11","MI420","MI423","BC07","SPVW12","BC14","MI425","MI419"),
      clustercol = c("mediumpurple","orchid1"),
      showlegend = F,
      legendmargin = c(2,2,2,2),
      showsp = F, 
      showdiv = T, 
      divsize = 0.5, 
      divcol = "black",  
      divtype = 1,
      legendkeysize = 10, 
      legendtextsize = 15, 
      linesize = 0.5, 
      pointsize = 4, 
      barbordercolour = "black",
      barbordersize = 0.2, 
      outputfilename = "../../visualization/admixture_figures/admixture_plots/neosho_admixed_lineages_2",
      imgtype = "pdf",
      imgoutput = "sep",
      height = 5, width = 30)

plotQ(admixed_3,
      showindlab = T, 
      useindlab=T,
      indlabsize = 8,
      indlabcol = "black",
      grplab = genomics_metadata_admixed,
      showgrplab = F,
      ordergrp = T,
      selgrp = "sample_id",
      subsetgrp =  c("NOIS18","NOIS08","NOIS07","BFORK30","BFORK24","SPRMO11","GLVR4","SPRMO19","CANEY15","AT12","AT05","AT08","AT09","AT02","AR29","AR30","ER33","ER35","ER20","ER30","ER44","ER17","BC01","BC08","BC09","BC12","SPVW14","SPVW07","MI422","MI421","SPVW02","SPVW05","SPVW11","MI420","MI423","BC07","SPVW12","BC14","MI425","MI419"),
      clustercol = c("lightgreen","mediumpurple","orchid1"),
      showlegend = F,
      legendmargin = c(2,2,2,2),
      showsp = F, 
      showdiv = T, 
      divsize = 0.5, 
      divcol = "black",  
      divtype = 1,
      legendkeysize = 10, 
      legendtextsize = 15, 
      linesize = 0.5, 
      pointsize = 4, 
      barbordercolour = "black",
      barbordersize = 0.2, 
      outputfilename = "../../visualization/admixture_figures/admixture_plots/neosho_admixed_lineages_3",
      imgtype = "pdf",
      imgoutput = "sep",
      height = 5, width = 30)

plotQ(admixed_4,
      showindlab = T, 
      useindlab=T,
      indlabsize = 8,
      indlabcol = "black",
      grplab = genomics_metadata_admixed,
      showgrplab = F,
      ordergrp = T,
      selgrp = "sample_id",
      subsetgrp =  c("NOIS18","NOIS08","NOIS07","BFORK30","BFORK24","SPRMO11","GLVR4","SPRMO19","CANEY15","AT12","AT05","AT08","AT09","AT02","AR29","AR30","ER33","ER35","ER20","ER30","ER44","ER17","BC01","BC08","BC09","BC12","SPVW14","SPVW07","MI422","MI421","SPVW02","SPVW05","SPVW11","MI420","MI423","BC07","SPVW12","BC14","MI425","MI419"),
      clustercol = c("lightgreen","orchid1","mediumpurple", "sienna4"),
      showlegend = F,
      legendmargin = c(2,2,2,2),
      showsp = F, 
      showdiv = T, 
      divsize = 0.5, 
      divcol = "black",  
      divtype = 1,
      legendkeysize = 10, 
      legendtextsize = 15, 
      linesize = 0.5, 
      pointsize = 4, 
      barbordercolour = "black",
      barbordersize = 0.2, 
      outputfilename = "../../visualization/admixture_figures/admixture_plots/neosho_admixed_lineages_4",
      imgtype = "pdf",
      imgoutput = "sep",
      height = 5, width = 30)
```

# CV Error
###All analyses were conducted in base R

## Cross-validation (10-fold) error results for ADMIXTURE analysesad
```{r}
#For each admixture analysis that I ran, I conducted a 10-fold cross validation to obtain error values for each value of K. The K value with the lowest cross-validation error is the K that is most likely given the data. Here I am graphing the cross-validation error values across all K for each analysis (hierarchical analysis for all samples, INCLUDING the hybrid, and hierarchial analysis for all samples, EXCLUDING the hybrid)

#Read in data for all samples INCLUDING the Hybrid (BFC10)
cv_all <- read_excel("../../raw_data/admixture_data/admixture_cv_data/cv_all.xlsx")
cv_pure <- read_excel("../../raw_data/admixture_data/admixture_cv_data/cv_pure.xlsx")
cv_admixed <- read_excel("../../raw_data/admixture_data/admixture_cv_data/cv_admixed.xlsx")

#Clean data for all samples, including the hybrid
cv_all <- as.data.frame(cv_all)
cv_pure <- as.data.frame(cv_pure)
cv_admixed <- as.data.frame(cv_admixed)
```

## Plot Cross Validation Results for each analysis 
```{r}
#cv error plot for all samples
cv_all_plot <- ggplot(cv_all, aes(x = k, y = cv)) + 
   geom_point(size = 3) + 
   geom_path(stat = "identity", size = 1) + 
   scale_x_continuous("K", labels = as.character(cv_all$k), breaks = cv_all$k) + 
   labs(x = "K", y = "10-Fold CV Error") + 
   theme_set(theme_cowplot()) + 
   theme(axis.text = element_text(size = 20)) + 
   theme(axis.title.x = element_text(face = "italic")) + 
   theme(axis.title = element_text(size = 20)) +
   theme(axis.title.x = element_blank())

#cv error plot for all samples, but ONLY PURE SAMPLES
cv_pure_plot <- ggplot(cv_pure, aes(x = k, y = cv_error)) + 
   geom_point(size = 3) + 
   geom_path(stat = "identity", size = 1) + 
   scale_x_continuous("K", labels = as.character(cv_pure$k), breaks = cv_pure$k) + 
   labs(x = "K", y = "10-Fold CV Error") + 
   theme_set(theme_cowplot()) + 
   theme(axis.text = element_text(size = 20)) + 
   theme(axis.title.x = element_text(face = "italic")) + 
   theme(axis.title = element_text(size = 20)) +
   theme(axis.title.x = element_blank())

#cv error plot for all samples, but ONLY ADMIXED SAMPLES
cv_admixed_plot <- ggplot(cv_admixed, aes(x = k, y = cv)) + 
   geom_point(size = 3) + 
   geom_path(stat = "identity", size = 1) + 
   scale_x_continuous("K", labels = as.character(cv_admixed$k), breaks = cv_admixed$k) + 
   labs(x = "K", y = "10-Fold CV Error") + 
   theme_set(theme_cowplot()) + 
   theme(axis.text = element_text(size = 20)) + 
   theme(axis.title.x = element_text(face = "italic")) + 
   theme(axis.title = element_text(size = 20))

pdf("/Users/joegunn/Desktop/Grad_School_Stuff/Research/Projects/Smallmouth_Bass_Genomics/visualization/admixture_figures/cv_plots/cv_error_plot_all_samples.pdf", width = 10, height = 5) 

ggplot(cv_all, aes(x = k, y = cv)) + 
   geom_point(size = 3) + 
   geom_path(stat = "identity", size = 1) + 
   scale_x_continuous("K", labels = as.character(cv_all$k), breaks = cv_all$k) + 
   labs(x = "K", y = "10-Fold CV Error") + 
   theme_set(theme_cowplot()) + 
   theme(axis.text = element_text(size = 20)) + 
   theme(axis.title.x = element_text(face = "italic")) + 
   theme(axis.title = element_text(size = 20)) +
   theme(panel.border = element_rect(colour = "black", fill=NA, size=1))

dev.off()

pdf("/Users/joegunn/Desktop/Grad_School_Stuff/Research/Projects/Smallmouth_Bass_Genomics/visualization/admixture_figures/cv_plots/cv_error_plot_all_pure_samples.pdf", width = 8, height = 4) 

ggplot(cv_pure, aes(x = k, y = cv_error)) + 
   geom_point(size = 3) + 
   geom_path(stat = "identity", size = 1) + 
   scale_x_continuous("K", labels = as.character(cv_pure$k), breaks = cv_pure$k) + 
   labs(x = "K", y = "10-Fold CV Error") + 
   theme_set(theme_cowplot()) + 
   theme(axis.text = element_text(size = 20)) + 
   theme(axis.title.x = element_text(face = "italic")) + 
   theme(axis.title = element_text(size = 20)) +
   theme(panel.border = element_rect(colour = "black", fill=NA, size=1))


dev.off()

pdf("/Users/joegunn/Desktop/Grad_School_Stuff/Research/Projects/Smallmouth_Bass_Genomics/visualization/admixture_figures/cv_plots/cv_error_plot_admixed_indivs.pdf", width = 5, height = 4) 

ggplot(cv_admixed, aes(x = k, y = cv)) + 
   geom_point(size = 3) + 
   geom_path(stat = "identity", size = 1) + 
   scale_x_continuous("K", labels = as.character(cv_admixed$k), breaks = cv_admixed$k) + 
   labs(x = "K", y = "10-Fold CV Error") + 
   theme_set(theme_cowplot()) + 
   theme(axis.text = element_text(size = 20)) + 
   theme(axis.title.x = element_text(face = "italic")) + 
   theme(axis.title = element_text(size = 20))

dev.off()
```
