---
title: 'Analysis 5: Admixture Mapping'
author: "Joe Gunn"
date: "2022-08-31"
output: html_document
---

# Project: Population genomic analysis of Smallmouth Bass and Neosho Bass in the Central Interior Highlands
We investigated the extent of genomic divergence, local directional selection, and admixture between the Smallmouth Bass (<i>Micropterus dolomieu</i>) and the Neosho Bass (<i>M. velox</i>) in the Central Interior Highlands (CIH) ecoregion of central north America. Specifically, we used ddRADseq data to assessed the phylogenomic relationship between and within species, characterizing inter- and intraspecific diversity and SNPs potentially under local directional selection at the population level. Additionally, we inferred the relative timing of admxiture in Neosho Bass streams where there is known introgressive hybridization with Smallmouth Bass to understand the influence of natural, historic geographic factors on mixing (stream capture or transient flooding) vs. anthropogenic factors (i.e., non-native introductions through stocking), which is known to have occurred widely in these economically valuable species. We ultimately hoped to provide novel insights into the diversity of endemic, ecologically important and popular sport fish in the CIH.

## Specific Aim: Admixture mapping analysis
In this analysis, we used the popgen.vcf data generated in Analysis 2 (SNP Filtering...) to assess the relative timing of admixture events between Smallmouth Bass and Neosho Bass. Specifically, we used moment statistics in MatLab with the software program MIXMAPPER to build a scaffold phylogeny with significantly pure (non-admixed) populations (based on our <i>a posteriori</i>) discovered populations in Analysis 4) and to map significantly admixed populations (also based on our discovered populations in Analysis 4) onto the tree. 

## Phases of analysis:
### Phase 1: Data preparation
### Phase 2: Admixture mapping analysis

## Libraries needed for analysis
```{r setup, echo = FALSE, include=FALSE}
library(readxl)
library(tidyverse)
library(cowplot)
library(devtools)
```

## PHASE 1: DATA PREPARATION
In this phase, we are converting the fully filtered, thinned dataset excluding Spotted Bass and the Spotted Bass hybrid (`admixture_phylogenomics_analysis/data/processed_vcf/03_popgen_spb.vcf`) into input files compatible with the MatLab software program MIXMAPPER. Specifically, we are converting the vcf file to .ind, .snp, and .geno files, which was done using the software program EIGENSOFT

All analyses conducted by calling commands in bash shell scripts were performed on the Lewis high-performance computing cluster at the University of Missouri <br>

<b>Programs needed</b>: <br><br>

MIXMAPPER v.2.0 (Lipson et al., 2013, 2014) <br>

<b>Citation</b>:<br>

Lipson, M., Loh, P. R., Levin, A., Reich, D., Patterson, N., & Berger, B. (2013). Efficient moment-based inference of admixture parameters and sources of gene flow. Molecular Biology and Evolution, 30, 1788–1802. doi:10.1093/molbev/mst099

Lipson, M., Loh, P. R., Patterson, N., Moorjani, P., Ko, Y. C., Stoneking, M., … Reich, D. (2014). Reconstructing Austronesian population history in Island Southeast Asia. Nature Communications, 5, 1–7. doi:10.1038/ncomms5689

PLINK v.1.90p (high-contig build; Change et al. 2015) <br>

<b>Citation</b>:<br>

Chang, C. C., Chow, C. C., Tellier, L. C. A. M., Vattikuti, S., Purcell, S. M., & Lee, J. J. (2015). Second-generation PLINK: Rising to the challenge of larger and richer datasets. GigaScience, 4, 1–16. doi:10.1186/s13742-015-0047-8

EIGENSOFT v.7.2.1 (Patterson, Price, & Reich, 2006; Price et al., 2006) <br>

<b>Citation</b>:<br>

Patterson, N., Price, A. L., & Reich, D. (2006). Population structure and eigenanalysis. PLoS Genetics, 2, 2074–2093. doi:10.1371/journal.pgen.0020190

Price, A. L., Patterson, N. J., Plenge, R. M., Weinblatt, M. E., Shadick, N. A., & Reich, D. (2006). Principal components analysis corrects for stratification in genome-wide association studies. Nature Genetics, 38, 904–909. doi:10.1038/ng1847

### STEP 1: Download the MIXMAPPER source code and the MIXMAPPER MCR distribution package.
In this step, we are downloading and installing the source code and MCR distribution package for the software program MIXMAPPER, which is written in C++ and MatLab. 

We have <i>NOT</i> included the downloaded source code, installed software, and compiled wrapper functions in this repository, so you will need to re-download and re-install these items. Follow Steps 1-2 below:

#### 1a: Navigate to the /code subdirectory and create a directory to hold the MIXMAPPER source code: `mkdir source_code`. Navigate inside this directory.

#### 1b: Go to the MIXMAPPER source code website: `http://cb.csail.mit.edu/cb/mixmapper/`. 

#### 1c: Navigate to the `source_code` directory: 

##### 1c.1. Download the source code (MixMapper_v2.0): `wget http://cb.csail.mit.edu/cb/mixmapper/MixMapper_v2.0.tar.gz`

###### 1c.1.1. Unzip the .gz file for the source code: `gunzip MixMapper_v2.0.tar.gz`

###### 1c.1.2. Unzip the .tar file for the source code: `tar -xvf MixMapper_v2.0.tar`

##### 1c.2: Download the MCR distribution package (MixMapperMCRdistrib_v1.02): `wget http://cb.csail.mit.edu/cb/mixmapper/MixMapperMCRdistrib_v1.02.tar.gz`

###### 1c.2.1. Unzip the .gz file for the MCR distribution package: `gunzip MixMapperMCRdistrib_v1.02.tar.gz`

###### 1c.2.2. Unzip the .tar file for the MCR distribution package: `tar -xvf MixMapperMCRdistrib_v1.02.tar`

There are two should be two folders within the `source_code` directory: `MixMapperMCRdistrib` and `MixMapper_v2.0`. 

###### 1c.2.3. Move both folders into a separate subdirectory called: `mixmapper_code`

### STEP 2: Install MCR runtime compiler (for running MATLAB if it is not installed on your local computer)
In this step, we are installing the MATLAB MCR runtime compiler.

#### 2a: Install the MCR runtime compiler; navigate into the `MixMapperMCRdistrib` directory.

##### 2a.1. Create a sub-directory: `mkdir MCR`. Navigate into this directory.

##### 2a.2. Inside the `MCR` directory, download the 64-bit MCR runtime compiler for your machine (I run on a MAC): `wget https://ssd.mathworks.com/supportfiles/MCR_Runtime/R2012a/MCR_R2012a_glnxa64_installer.zip`

Note: in the 'instructions.txt' file provided in the MixMapperMCRdistrib folder, the link address for the runtime compiler is deprecated and will not take you to the correct site. Use the link address given above in Step 2a.2. Verify that the .zip file is named the same as in the 'instructions.txt' file ('MCR_R2012a_glnxa64_installer.zip)

##### 2a.3. Decompress the .zip file: `unzip MCR_R2012a_glnxa64_installer.zip`
Decompressing this .zip file will unpack several sub-folders, command line software executables, and instruction text files.

###### 2a.3.1. Navigate back out to the `MixMapperMCRdistrib` directory, and open the 'instructions.txt' file. Follow lines 13 - 21 exactly. Set the path for the installation to the `MCR` directory

##### 2a.4. Once all necessary lines of code have been uncommented (from 'instructions.txt', lines 13 - 21), install the MCR runtime compiler: `./install -inputFile installer_input_mod.txt`
Installing the runtime compiler will generate a new subdirectory within the `MCR` directory called `v717`. This is where the runtime compiler is and where paths for MixMapper need to be set.

### STEP 3: Compile necessary wrapper functions for MIXMAPPER.
In this step, we are compiling two source files, which will be converted to executable code for running MIXMAPPER. These source files are: 'compute_moment_stats.cpp' and 'compute_most_additive_trees.cpp'

#### 3a. Navigate to the `MixMapper_v2.0` directory; make sure the two files needing compilation are in the folder (see STEP 3 description above)

##### 3a.1. Compile 'compute_moment_stats.cpp'; in the command line, start an interaction session and run the bash command: `srun g++ --02 compute_moment_stats.cpp -o compute_moment_stats`
Running this command will generate an executable for compute_moment_stats

##### 3a.2. Compile 'compute_most_additive_trees.cpp'; in the command line, start an interaction session and run the bash command: `srun g++ --02 compute_most_additive_trees.cpp -o compute_most_additive_trees`
Running this command will generate an executable for compute_most_additive trees.

### STEP 4: Generate input files for MIXMAPPER.
In this step, we are using the fully filtered, non-thinned VCF dataset excluding the spotted bass hybrid `admixture_phylogenomics_analysis/data/processed_vcf/popgen_spb_hybrid.vcf` generated in Analysis 2 (SNP Filtering...) and generating input files for MIXMAPPER. 

#### 4a: Convert the fully filtered, non-thinned VCF dataset into PED (pedigree) format.
In this step, we are using the software program PLINK to convert the full dataset to .ped format, which will then be used in subsequent steps to convert to eigenstrat format.

##### 4a.1. Generate a general bash shell script header for all PLINK commands
In this step, we are generating a universal bash shell script to run all file type conversions in PLINK. Here, we only generate the annotated header for the shell script, and in subsequent steps, we include and provide details on each file conversion and the associated lines of code in Rmd chunks. The shell script for running all PLINK code is called "plink.sh".

###### 4a.1.1: Copy and paste the code below in a new shell script file:

##### Generate PLINK bash shell script: `code/shell_scripts/plink.sh`
```{bash}
#! /bin/bash

#SBATCH -p Lewis  # use the Lewis partition
#SBATCH -J get_ped # give the job a custom name
#SBATCH -o plink.out  # give the job output a custom name
#SBATCH -t 0-01:00  # two hour time limit

#SBATCH -N 1  # number of nodes
#SBATCH -n 1  # number of cores (AKA tasks)

## Commands here run only on the first core

## Modules to load

# module load rss/rss-2020
# module load plink/plink-high-contig-1.90p (UNCOMMENT THIS LINE TO RUN THE CODE)
```

Make sure to copy the `admixture_phylogenomics_analysis/data/processed_vcf/03_popgen_spb.vcf` dataset generated in Analysis 2 into a new sub-directory of the `data/` folder in this directory: `data/processed_vcf`

##### 4a.2. Copy and paste the code below in the shell file generated in Step 4a.1.1 above, which will generate a .ped file for the full dataset:

##### Command line code for geenrating .ped. UNCOMMENT this code in the shell script:
```{bash}
#plink --vcf ../../data/processed_vcf/01_popgen_spb_hybrid.vcf --recode --recode-INFO-all --out ../../data/processed_ped/popgen
```

Run: `sbatch plink.sh`

This code generates the file: `data/processed_bed/all_samples/popgen.ped`, along with three other accompanying files, which should be used in subsquent file converstion steps:

<b>Files</b>: <br><br>
1) popgen.log
2) popgen.map
3) popgen.nosex

#### 4b: Convert the .ped file to EIGENSFOT format (.eigenstratgeno, .snp, .ind).
In this step, we are using the software program EIGENSOFT to convert the .ped format file generated in step 4a above into .eigenstratgeno, .snp, and .ind files (collectively eigenstrat files) which will then be used as input in MIXMAPPER. 

Specifically, we are using the `convertf` command to convert .ped to .eigenstrat

##### 4b.1. Modify the `popgen.map` file generated in Step 4a above.
In this step, we need to modify the `popgen.map` file generated in Step 4a above, because the software program EIGENSOFT expects to see chromosome and SNP designations in a particular format. The software was built to analyze data sequenced from humans and aligned to the human genome. Thus, the "chromosome" column (column 1) of the .map file will only recognize autosomal chrosomosome designations (1-22). Column 2 of the .map file contains SNP identifier information, and this can be in any format.

Additionally, EIGENSOFT expects input in columns 3 and 4 (genetic and physical distance, respectively) that is not zero. These values are based on physical distances between each SNP along chromosomes and genetic distances in cM, information that we don't have for these data. The program MIXMAPPER requires this information to make inferences about the timinig of admixture, and as such it is required as input. However, this information is not needed for MIXMAPPER in principle to make inferences about significantly admixed populations vs. pure or incompletely sorted populations, which is our aim in this analysis. So, we generated arbitrary, equal numerical values for these data (d) based arbitrarily on values of x (the numerical value of the RAD kmer) and y (nucleotide position on RAD tags, which is given in our .vcf file), using the custom formula:

d = (100x-y)/(100)

This equation is applied below in Step 4b.1.1 to modify the popgen.map file.

###### 4b.1.1. Read in and modify the `popgen.map` file; run the Rmd chunk below:

##### Read in and modify file:
```{r}
# Read in the map file and designate a temporary header column
map_file <- read_tsv("data/processed_ped/popgen.map", 
                     col_names = c("chrom","snp","gen_dist","pos"))

# Omit the empty second column
map_file <- map_file[,-c(2)]

# Rename "chrom" column as "snp"
colnames(map_file) <- c("snp","gen_dist","pos")

# Generate a column of "1" as a placeholder for chromosome designation, to add as the first column to the dataframe
chrom <- c(rep("1",times = 50828))

# combine "chrom" column and the rest of the data
map_file <- cbind(chrom, map_file)

# split each SNP name into component pieces to isolate the numerical value of the SNP
map_file_str_split <- map_file %>%
  separate(col = snp, into = c("rad","kmer","num"), sep = "_", remove = F)

# Omit leading zeros from the SNP identifier number ("num")
map_file_str_split$num <- sub("^0+", "", map_file_str_split$num) 

# Make sure the "num" (number of the RAD kmer designation) is a numeric value
map_file_str_split <- map_file_str_split %>%
  mutate(num = as.numeric(num))

# Run the equation to generate column genetic distance (gen_dist) 
map_file_str_split <- map_file_str_split %>%
  mutate(gen_dist = ((100*num) - pos)/100)

# Round the genetic distance value to the nearest integer
map_file_str_split <- map_file_str_split %>%
  mutate(gen_dist = round(gen_dist, digits = 0))

# Create an additional column for physical distance (phys_dist) and set equal to gen_dist
map_file_str_split <- map_file_str_split %>%
  mutate(phys_dist = gen_dist)

map_file <- map_file_str_split %>%
  select(chrom, snp, gen_dist, phys_dist)

# Save the new .map file
write_tsv(map_file, 
          file = "data/processed_ped/popgen_mod.map", 
          col_names = FALSE)
```

The modified popgen.map file is saved as `data/processed_ped/popgen_mod.map` to distinguish it from the original file name. 

##### 4b.2. Modify the `popgen.ped` file generated in Step 4a above.
In this step, we need to modify the `popgen.ped` file generated in Step 4a above, because the software program EIGENSOFT expects

###### 4b.2.1. Navigate to the directory where the `popgen.ped` file is located

###### 4b.2.2. Edit column 2 (Individual Name) so that it has the sample name as given in column 1; copy and paste the following code on the command line and run:
```{bash}
#awk '{OFS=" ";if($2=="1") {$2=$1;};print;}' popgen.ped > popgen_mod.ped
```

###### 4b.2.3. Edit column 6 (Phenotype) so that each entry is "1"; copy and paste the following on the command line and run:
```{bash}
#awk '{OFS=" ";if($6=="-9") {$6="1";};print;}' popgen.ped > popgen_mod.ped
```

This step must be taken even though we technically do not have a phenotype that we are interested in. If column 6 has a -9, it will be ignored. Since this is every sample, the whole file will be ignored.

##### 4b.3. Generate par.PED.EIGENSTRAT file as input for EIGENSOFT
In this step, we are creating a specialized file type ('par.PED.EIGENSTRAT'), which is used as input for the -p flag with the convertf command in EIGENSOFT. This file reads .ped, .map, and .ind files as input and generates .eigenstratgeno, .snp, and .ind files as output

###### 4b.3.1. Navigate into the `data/batch_cmd_lists/` folder and copy and paste the code below in a new nano .txt file:

##### Generate par.PED.EIGENSTRAT input file: `code/batch_cmd_lists/par.PED.EIGENSTRAT`. UNCOMMENT each line of this file.
```{bash}
#genotypename:    ../../data/processed_ped/popgen_mod.ped
#snpname:         ../../data/processed_ped/popgen_mod.map
#indivname:	 ../../data/processed_ped/popgen_mod.ped
#outputformat:    EIGENSTRAT
#genotypeoutname: ../../data/processed_eigensoft/popgen.eigenstratgeno
#snpoutname:	 ../../data/processed_eigensoft/popgen.snp
#indivoutname:    ../../data/processed_eigensoft/popgen.ind
#familynames:     NO
```

##### 4b.4. Generate a general bash shell script header for all EIGENSOFT commands
In this step, we are generating a universal bash shell script to run all file type conversions in EIGENSOFT. Here, we only generate the annotated header for the shell script, and in subsequent steps, we include and provide details on each file conversion and the associated lines of code in Rmd chunks. The shell script for running all EIGENSOFT code is called "eigensoft.sh".

###### 4b.4.1: Copy and paste the code below in a new shell script file:

##### Generate EIGENSOFT bash shell script: `code/shell_scripts/eigensoft.sh`
```{bash}
#! /bin/bash

#SBATCH -p Lewis  # use the Lewis partition
#SBATCH -J eigensoft # give the job a custom name
#SBATCH -o eigensoft.out  # give the job output a custom name
#SBATCH -t 0-01:00  # two hour time limit

#SBATCH -N 1  # number of nodes
#SBATCH -n 1  # number of cores (AKA tasks)

# Commands here run only on the first core

## Modules to load

# module load rss/rss-2020
# module load eigensoft/eigensoft-7.2.1
```

##### 4b.5. Convert PED files to EIGENSTRAT format.

###### 4b.5.1. Copy and paste the code below in the shell file generated in Step 4b.1.1 above, which will generate .eigenstratgeno, .snp, and .ind files for the full dataset:

##### Command line code for generating .eigenstrat. UNCOMMENT this code in the shell script:
```{bash}
# convertf -p ../batch_cmd_lists/par.PED.EIGENSTRAT
```

Run: `sbatch eigensoft.sh`

This code generates three files here: `data/processed_eigensfot/`; these files include:

<b>Files</b>: <br><br>
1) popgen.ind
2) popgen.snp
3) popgen.eigenstratgeno

#### 4c: Modify .ind file to include information about populations inferred in Analysis 4 (population inference).
In this step, we are modifying the `popgen.ind` file generated by EIGENSOFT to include information about the population designations for each sample as inferred in the population inference analysis. We are appending the population designations from the `metadata_populations.xlsx` file generated in Phase 4 of Analysis 4. 

##### 4c.1. Modify .ind file to include population designations for all samples; run the Rmd chunk below:

##### Modify the .ind file:
```{r}
# Read in the map file and designate a temporary header column
ind_file <- read_tsv("data/processed_eigensoft/popgen.ind",
                     col_names = c("sample_id"))

ind_file <- ind_file %>%
  separate(sample_id, into = c("sample_id","u"), sep = " ", remove = T)

# Load in metadata for populations generated in Analysis 4 (population analysis)
load("../population_analysis/data/metadata/metadata_population.Rda")

# cbind ind_file with population designations
ind_file <- cbind(ind_file, metadata_populations)

# omit column 3
ind_file <- ind_file[,-c(3)]

# Save the new .ind file
write_tsv(ind_file, 
          file = "data/processed_eigensoft/popgen_mod.ind",
          col_names = FALSE)
```

### ----------------------- END OF PHASE 1: DATA PREPARATION  ----------------------- ###

## PHASE 2: ADMIXTURE MAPPING ANALYSIS 
In this phase of the analysis, we are using the software program MIXMAPPER (see Phase 1 for information on programs needed and associated citations) to infer significantly admixed and pure populations based on population inference in Analysis 4. We are then mapping any inferred admixed populations onto a scaffold tree of pure populations to determine the parents of origin as well as relative timing of admixture events among populations.

### STEP 1: Infer admixed and pure populations
In this step, we are running preliminary analyses in MIXMAPPER to determine which populations in our dataset are 1) genetically pure and genetically admixed using the `compute_moment_statistics` wrapper function previously compiled in Phase 1; and 2) determine the most additive possible phylogenetic relationship among populations based on moment statistics (linear combinations of allele frequencies) using the `compute_most_additive_trees` wrapper function.

#### 1a: Compute moment statistics

##### 1a.1. Navigate to the `code/source_code/mixmapper_code/MixMapper_v2.0/` directory. Copy and paste the following code and run on the command line:
We are running the command line code with the following specs: <br>

n: no re-sampling of individuals <br>
1000 bootstrap replicates <br>
bootstraping over 50 cM blocks <br>

##### Command line code for computing moment statistics: UNCOMMENT this code.
```{bash}
#./compute_moment_stats ../../../../data/processed_eigensoft/popgen_mod.ind ../../../../data/processed_eigensoft/popgen.snp ../../../../data/processed_eigensoft/popgen.eigenstratgeno ../../../../data/mixmapper_data/popgen n 1000 50
```

The command line should update in real-time with progress of bootstrap replication completion. This code generates 5 output files, which are used as summary data and as input for the next analysis step: <br>

1) `popgen.f2_boots.txt` <br>
2) `popgen.f2.tab` <br>
3) `popgen.h_boots.txt` <br>
4) `popgen.neg_f3.txt` <br>
5) `popgen.pops.txt` <br>

#### 1a.2. Generate a list of inferred non-admixed populations; copy the `popgen.pops.txt` to a new file: `popgen_scaffold_pops.txt`. Remove admixed populations inferred in Step 1a above.
 
#### 1b: Compute most additive trees

##### 1b.1. Navigate to the `code/source_code/mixmapper_code/MixMapper_v2.0/` directory. Copy and paste the following code and run on the command line:
We are running the command line code with the following specs: <br>

10000 data subsets <br>
selecting from pure populations detected by compute_moment_stats <br>

##### Command line code for computing most additive trees: UNCOMMENT this code.
```{bash}
#./compute_most_additive_trees ../../../../data/mixmapper_data/popgen.f2.tab 10000 ../../../../data/mixmapper_data/popgen_scaffold_pops.txt > popgen_trees.out
```

### STEP 2: Admixture mapping: infer admixture origins and build admixture trees with MIXMAPPER.
In this step, we are using the data generated from the `compute_moment_stats` and `compute_most_additive_trees` functions in Step 1 to place admixed populations (ELK, UPPARK, BAYOU, AND ILLI) on the scaffold tree. We are using the MCR runtime distribution; it is fifrst necessary to locate the `/v717` folder within the MCR runtime distributor to include as a path name in the MIXMAPPER wrapper function.

#### 2a: Get path for `/v717/` folder in MCR runtime distributor; navigate to `/MixMapperMCRdistrib/MCR/v717/` and copy the path.

#### 2b: Navigate to the top level of `/MixMapperMCRdistrib/`. nano into the `cmd.txt` folder, which is the editable input file for admixture fitting in MIXMAPPER.

##### 2b.1. Modify the `cmd.txt` file; copy and paste the Rmd chunk below into the txt file.
In the "pop_data_reps" field, we included a full path to all output data from compute_moment_stats, which is read by MIXMAPPER. In the "scaffold_pop_names = " field, we input the names of each of our non-admixed populations to use as a scaffold tree for placement of admixed populations. In the admixture fitting fields, we set 100 bootstrap replicates and specified the program to compute two-way admixtures. 

##### Modify file: `cmd.txt`: UNCOMMENT 
```{r}
#% read input data generated by compute_moment_stats (C++ code)
#% note '~' doesn't seem to work in path?
#pop_data_reps = read_MixMapper_input('../../../../data/mixmapper_data/popgen');

#% choose populations for scaffold tree
#scaffold_pop_names = {'LMULB' 'MIDARK' 'MISS' 'SKIA' 'WHITE'};

#% run bootstrap reps in parallel
#% matlabpool open 4

#options.branch_sets = 1;
#% fit ELK as two-way admixture
#[trees,fits] = MixMapper(pop_data_reps(1:100),scaffold_pop_names,'ELK','',options);

#% fit ILLI as two-way admixture
#[trees,fits] = MixMapper(pop_data_reps(1:100),scaffold_pop_names,'ILLI','',options);

#% fit UPPARK as two-way admixture
#[trees,fits] = MixMapper(pop_data_reps(1:100),scaffold_pop_names,'UPPARK','',options);

#% fit BAYOU as two-way admixture
#[trees,fits] = MixMapper(pop_data_reps(1:100),scaffold_pop_names,'BAYOU','',options);

#% write tree from all-data run in Newick format
#phytreewrite('tree1.tree',trees(1));
```

#### 2c: Run MIXMAPPER; copy and paste the code below and run on the command line within the directory containing the MCR distributor:

##### Command line code for running MIXMAPPER: UNCOMMENT before running.
```{r}
# ./run_MixMapper_wrapper.sh MCR/v717 cmd.txt
```

This code will show real-time progress of running through each replicate. It will produce an output table of all admixture mappings (which is used as the basis for maintext Table 2 of our manuscript) as well as a phylogenetic tree .tree file (which was manipulated further in Excel and used as the basis for maintext Figure 4 of our manuscript).

### ----------------------- END OF PHASE 2: ADMIXTURE MAPPING ANALYSIS  ----------------------- ###

