---
title: "Analysis 2: SNP Filtering, Data Processing, and Preliminary Calculations"
author: "Joe Gunn"
date: "12/11/2019"
output: html_document
---

# Project: Population genomic analysis of Smallmouth Bass and Neosho Bass in the Central Interior Highlands
We investigated the extent of genomic divergence, local directional selection, and admixture between the Smallmouth Bass (<i>Micropterus dolomieu</i>) and the Neosho Bass (<i>M. velox</i>) in the Central Interior Highlands (CIH) ecoregion of central north America. Specifically, we used ddRADseq data to assessed the phylogenomic relationship between and within species, characterizing inter- and intraspecific diversity and SNPs potentially under local directional selection at the population level. Additionally, we inferred the relative timing of admxiture in Neosho Bass streams where there is known introgressive hybridization with Smallmouth Bass to understand the influence of natural, historic geographic factors on mixing (stream capture or transient flooding) vs. anthropogenic factors (i.e., non-native introductions through stocking), which is known to have occurred widely in these economically valuable species. We ultimately hoped to provide novel insights into the diversity of endemic, ecologically important and popular sport fish in the CIH.

## Specific Aim: Filtering SNP and individual sample data and performing preliminary allele frequency calculations
In this analysis, we filter the raw VCF data, representing 95 samples consisting of Spotted Bass, Smallmouth Bass, and Neosho Bass, and 127,429 ddRADseq loci markers for all samples. We performed the following filtering steps, in this order:

<b> Filtering Steps for full dataset </b>: <br>
1. removal of samples with greater than 20% missing genotype calls across SNPs <br>
2. removal of SNPs with sequencing depth less than 15 X <br>
3. removal of SNPs with phred quality score less than 20 <br> 

At this point in the filtering process, we split the data into two distinct datasets: 1) a dataset for population genomic analyses ("popgen"; dataset A), and 2) a dataset for population delimitation using haplotype inference in the software program fineRADstructure ("finerad"; dataset B), which utilizes multiple SNPs per RAD kmer in ddRADseq data to infer coancestry matrices between all pairs of samples. We performed the following filtering steps for each datatset separately:

<b> Filtering Steps for the POPGEN dataset (dataset A) </b>: <br>
1. removal of all but one SNP per RAD kmer <br>
2. removal of all SNPs with minor allele count threshold less than 2 (minor allele frequency ~0.011) <br>
3. removal of all SNPs with heterozygosity greater than 45% <br>

<b> Filtering Steps for the FINERAD dataset (dataset B) </b>: <br>
1. removal of all SNPs with minor allele count threshold less than 2 (minor allele frequency ~0.011) <br>
2. removal of all SNPs with heterozygosity greater than 45% <br>

We detail each step of the filtering process below.

## Phases of analysis:
### Phase 1: Filter raw VCF file

## Libraries needed for Analysis
```{r libraries, echo = FALSE}
library(readxl)
library(tidyverse)
library(cowplot)
library(devtools)
```

## PHASE 1: Filter raw VCF file
In this phase of the analysis, we are filtering the final, raw VCF file which was generated with stringent filtering criteria at Floragenex, Inc. (see `project_info/Floragenex_project_report.pdf`). The file was originally named "AR21_Aligned_Genotypes_stringent.vcf"; we renamed the file "smb_genomics.vcf" for convenience, and the file can be found here: `raw_data/smb_genomics.vcf`. <br>

Importantly, raw sequences (.fastq files) generated by Illumina ddRAD-seq were NOT mapped to a reference genome, as there was not a closely related and annotated reference genome at the time of analysis. Thus, the first column of the VCF ("#CHROM"), which customarily contains chromosomal information, instead includes a numerical designation for the RAD kmer (95 bp haplotype sequence) on which the SNP was found. The nucleotide position of the SNP ("#POS" column) is also given. Each RAD kmer is also referred to as a "contig". <br>

All analyses conducted by calling commands in bash shell scripts were performed on the Lewis high-performance computing cluster at the University of Missouri <br>

<b>See below for information on SNPs in the original VCF file</b>: <br>
<b>Total SNPs:</b> 127,429

<b>Programs needed</b>: <br>

VCFTools v.0.1.14 (Danecek et al. 2011) <br>

<b>Citation</b>:<br>

Danecek, P., Auton, A., Abecasis, G., Albers, C. A., Banks, E., DePristo, M. A., … Durbin, R. (2011). The variant call format and VCFtools. Bioinformatics, 27, 2156–2158. doi:10.1093/bioinformatics/btr330 <br>


GATK v.4.0.1.1 (Depristo et al. 2011)

<b>Citation</b>:<br>

Depristo, M. A., Banks, E., Poplin, R., Garimella, K. V., Maguire, J. R., Hartl, C., … Daly, M. J. (2011). A framework for variation discovery and genotyping using next-generation DNA sequencing data. Nature Genetics, 43, 491–501. doi:10.1038/ng.806 <br><br>

### STEP 1: Prepare the VCF file for all downstream analyses.
In this step, we are copying the raw VCF file containing SNPs called with stringent filters at Floragenex, Inc. (AR21_Aligned_genotypes_stringent.vcf) and renaming for downstream analyses. Additionally, we are deleting one row of the header in the file: "##contig=NA" because this header line is for some reason incompatible with the snp table generation tool in GATK.

#### 1a: Save the edited VCF file as `raw_data/smb_genomics.vcf`

### STEP 2: Generate a general bash shell script header for all VCF commands with which to run all VCF commands
In this step, we are generating a universal bash shell script to run all SNP and individual filtering on the full VCF file with the program VCFTools. Here, we only generate the annotated header for the shell script, and in subsequent steps, we include and provide details on each filtering decision and the associated lines of code in Rmd chunks. The shell script for running all VCFTools is called "vcftools.sh".

#### 2a: Generate the header for a general bash shell script to run VCFTOOLS:

##### Generate vcftools bash shell script: `code/shell_scripts/vcftools.sh`
```{bash}
#! /bin/bash

#SBATCH -p Lewis  # use the Lewis partition
#SBATCH -J get_meanDP  # give the job a custom name
#SBATCH -o vcf_meanDP-%j.out  # give the job output a custom name
#SBATCH -t 0-02:00  # two hour time limit

#SBATCH -N 1  # number of nodes
#SBATCH -n 1  # number of cores (AKA tasks)

## Load VCFtools module
#module load rss/rss-2020 (UNCOMMENT THIS LINE TO RUN THE CODE)
#module load vcftools/vcftools-v0.1.14 (UNCOMMENT THIS LINE TO RUN THE CODE)
```

### STEP 3: Filter out individual samples with greater than 20% missing data across all SNPs.
In this step, we are filtering the dataset to omit any individual samples that did not genotype well across SNPs. Specifically, we are omitting all individuals that had greater than 20% missing genotype data (more than 20% of SNPs had no genotype call).

#### 3a: Copy and paste the code below in the shell file generated in Step 2a above, which will generate a file with missing genotype call rate information for all samples:

##### Command line code for obtaining missing genotype call rate information. UNCOMMENT this code in the shell script:
```{bash}
#vcftools --vcf ../../../raw_data/smb_genomics.vcf --missing-indv --out ../../data/filtering_data/smb_missing
```

Run: `sbatch vcftools.sh`

This code generates the file: `data/filtering_data/smb_missing.imiss`. This file should be used to generate a separate list of sample IDs with missing genotype call rates greater than 20%, which will subsequently be used to omit these samples from the full VCF.

#### 3b: Generate list of sample IDs for individuals with greater than 20% missing data; run the Rmd chunk below:

##### Generate list of sample IDs: `smb_genomics_badsamples.txt`
```{r}
## read in .imiss file
smb_genomics_missing <- read_tsv("data/filtering_data/smb_genomics_missing.imiss")

## Convert to data frame and mutate individual to factor
smb_genomics_missing <- as.data.frame(smb_genomics_missing) %>%
  mutate(INDV = factor(INDV))

## Generate list of samples above missing genotype threshold
smb_genomics_badsamples <- smb_genomics_missing %>%
  filter(F_MISS > 0.2)

# Select the first column and convert to data frame
smb_genomics_badsamples <- data.frame(smb_genomics_badsamples[,c(1)]) 

# Save the list of bad samples as a .txt file without column names
write_tsv(smb_genomics_badsamples, 
          file = "data/filtering_data/smb_genomics_badsamples.txt", 
          col_names = FALSE)
```

#### 3c: Generate a bar plot showing the missing genotype proportions of all samples; run the Rmd chunk below:

##### Generate bar plot for badsamples: `figures/filtering_processing_figures/badsamples.pdf`
```{r}
## Remove the "_1" string from each sample name
smb_genomics_missing <- as.data.frame(smb_genomics_missing) %>%
  mutate(INDV = factor(INDV))  %>%
  separate(col = INDV, into = c("INDV", NA), sep = "_1")

# Get only columns 1 and 2
smb_genomics_missing <- smb_genomics_missing %>%
  select(INDV, F_MISS)

## Plot figure
pdf("figures/filtering_processing_figures/badsamples.pdf", width=7, height=25)

ggplot(smb_genomics_missing, aes(x = INDV, y = F_MISS)) + 
  geom_bar(position=position_dodge(), stat="identity", show.legend = T, fill = "grey", color = "black") + 
  ylim(0,1) +
  geom_hline(yintercept = 0.2, linetype = "longdash", color = "red", size = 2) + 
  theme_set(theme_cowplot(12)) + 
  labs(x = "Sample ID", y = "Proportion of Missing Data") + 
  coord_flip() + 
  theme(axis.text.x = element_text(angle = 60, hjust = 1)) + 
  theme(axis.title = element_text(size = 25)) + 
  theme(axis.text = element_text(size = 20)) +
  theme(panel.border = element_rect(colour = "black", fill=NA, size=1))

dev.off()
```

This plot corresponds to Supplementary Figure S1 in the manuscript.

#### 3d: Copy and paste the code below in the shell file generated in Step 2a above, which will eliminate the poor quality samples discovered in Steps 3a - 3c above:

##### Command line code for omitting bad samples. UNCOMMENT this code in the shell script:
```{bash}
#vcftools --vcf ../../../raw_data/smb_genomics_edited.vcf --remove ../../data/filtering_data/smb_genomics_badsamples.txt --recode --recode-INFO-all --out ../../data/processed_vcf/01_filter_badsamples
```

Run: `sbatch vcftools.sh`

##### Important Note: the generated file will have an additional extension (.recode), which is not necessary for downstream analyses. Every time a new VCF file was filtered, we manually removed this piece of the extension by copying the file to a new name (see below) and deleting the .recode.vcf file.

This code generates the file: `data/processed_vcf/01_filter_badsamples.vcf`. This file serves as input for the next filtering step.

<b>Filtering results</b>: <br><br>
<b>Number of individuals filtered out</b> = 3 <br>
<b>Individuals filtered</b>: BFORK32 (Neosho), GRSPB34 (Spotted), ER05 (Neosho) <br>
<b>Individuals remaining</b> = 92 <br>

### STEP 4: Filter SNPs for a minimum depth of coverage of 15X
In this step, we are filtering the dataset generated in Step 3d above to retain only SNPs that have a minimum depth of coverage across all individuals of 15X. 

#### 4a: Copy and paste the code below in the shell file generated in Step 2a above, which will eliminate SNPs with depth of coverage less than 15X:

##### Command line code for omitting low depth SNPs. UNCOMMENT this code in the shell script:
```{bash}
#vcftools --vcf ../../data/processed_vcf/01_filter_badsamples.vcf --minDP 15 --recode --recode-INFO-all --out ../../data/processed_vcf/02_filter_depth
```

Run: `sbatch vcftools.sh`

This code generates the file: `data/processed_vcf/02_filter_depth.vcf`. This file serves as input for the next filtering step.

<b>Filtering results</b>: <br><br>
<b>SNPs filtered out</b> = 0 <br>
<b>SNPs retained </b> = 127,429 <br>

### STEP 5: Filter SNPs for a minimum depth of coverage of 15X
In this step, we are filtering the dataset generated in Step 4a above to retain only SNPs that have a minimum depth of coverage across all individuals of 15X. 

#### 5a: Copy and paste the code below in the shell file generated in Step 2a above, which will eliminate SNPs with phred quality score less than 20.

##### Command line code for omitting poor quality SNPs. UNCOMMENT this code in the shell script:
```{bash}
#vcftools --vcf ../../data/processed_vcf/02_filter_depth.vcf --minQ 20 --recode --recode-INFO-all --out ../../data/processed_vcf/03_filter_qual
```

Run: `sbatch vcftools.sh`

This code generates the file: `data/processed_vcf/03_filter_qual.vcf`. This file serves as input for the next filtering step.

<b>Filtering results</b>: <br><br>
<b>SNPs filtered out</b> = 406 <br>
<b>SNPs retained </b> = 127,023 <br>

### STEP 6: Filter SNPs for minimum genotype call rate across individuals of 80%
In this step, we are filtering the dataset generated in Step 5a above to retain only SNPs that have a minimum genotype call rate across individual samples of 80%. Any SNPs genotyped at less than 80% of SNPs are removed from further analyses.

#### 6a: Copy and paste the code below in the shell file generated in Step 2a above, which will eliminate SNPs with genotype call rates across individuals less than 80%.

##### Command line code for omitting poor quality SNPs. UNCOMMENT this code in the shell script:
```{bash}
#vcftools --vcf ../../data/processed_vcf/03_filter_qual.vcf --minQ 20 --recode --recode-INFO-all --out ../../data/processed_vcf/04_filter_genotype
```

Run: `sbatch vcftools.sh`

This code generates the file: `data/processed_vcf/04_filter_genotype.vcf`. This file serves as input for the next filtering step.

<b>Filtering results</b>: <br><br>
<b>SNPs filtered out</b> = 0 <br>
<b>SNPs retained </b> = 127,023 <br>

### STEP 7: Split the dataset into two datasets for two avenues of analysis and filter each dataset further
In this step, we split the dataset filtered in Step 6a into two distinct datasets to be used in specific downstream analyses: 1) a full, <i><b>thinned</b></i> dataset retaining only one SNP per RAD tag (to avoid bias in population genomic analyses from tight nucleotide linkage), and 2) a full, non-thinned dataset retaining all SNPs to be used in haplotype inference of fine population structure. Both of these datasets are further filtered for a minimum minor allele count and maximum heterozygosity threshold, which are detailed in the steps below.

#### 7a: Generate POPGEN dataset. 

##### 7a.1. Thin popgen dataset to retain only one SNP per RAD kmer.

###### 7a.1.1. Copy and paste the code below in the shell file generated in Step 2a above, which will retain only one SNP per RAD kmer in a thinning procedure (thin 100 used since RAD tags are each 95 base pairs long)

##### Command line code for one SNP per RAD tag. UNCOMMENT this code in the shell script:
```{bash}
#vcftools --vcf ../../data/processed_vcf/04_filter_genotype.vcf --thin 100 --recode --recode-INFO-all --out ../../data/processed_vcf/A_01_popgen_filter_thin
```

Run: `sbatch vcftools.sh`

This code generates the file: `data/processed_vcf/A_01_popgen_filter_thin.vcf`. This file serves as input for the next filtering step for the POPGEN (thinned) dataset.

<b>Filtering results for POPGEN (thinned) dataset </b>: <br><br>
<b>SNPs filtered out</b> = 61,769 <br>
<b>SNPs retained </b> = 65,254 <br>

##### 7a.2. From the POPGEN dataset, filter out SNPs with minor allele frequency less than or equal to 2 (minor allele frequency ~0.011).

##### 7a.2.1. Copy and paste the code below in the shell file generated in Step 2a above, which will filter out any SNPs with minor allele count less than or equal to 2 (this is equivalent to a minor allele frequency of ~0.011).

#### Command line code for minor allele count filtering. UNCOMMENT this code in the shell script:
```{bash}
#vcftools --vcf ../../data/processed_vcf/A_01_popgen_filter_thin.vcf --mac 2 --recode --recode-INFO-all --out ../../data/processed_vcf/A_02_popgen_filter_mac
```

Run: `sbatch vcftools.sh`

This code generates the file: `data/processed_vcf/A_02_popgen_filter_mac.vcf`. This file serves as input for the next filtering step for the POPGEN (thinned) dataset.

<b>Filtering results for POPGEN (thinned) dataset </b>: <br><br>
<b>SNPs filtered out</b> = 13,492 <br>
<b>SNPs retained </b> = 51,762 <br>

#### 7b: Generate FINERAD dataset. 

##### 7b.1. From the FINERAD dataset, filter out SNPs with minor allele frequency less than or equal to 2 (minor allele frequency ~0.011).

###### 7b.1.1. Copy and paste the code below in the shell file generated in Step 2a above, which will filter out any SNPs with minor allele count less than or equal to 2 (this is equivalent to a minor allele frequency of ~0.011).

#### Command line code for minor allele count filtering. UNCOMMENT this code in the shell script:
```{bash}
#vcftools --vcf ../../data/processed_vcf/04_filter_genotype.vcf --mac 2 --recode --recode-INFO-all --out ../../data/processed_vcf/B_01_finerad_filter_mac
```

Run: `sbatch vcftools.sh`

This code generates the file: `data/processed_vcf/B_01_finerad_filter_mac.vcf`. This file serves as input for the next filtering step for the FINERAD (non-thinned) dataset

<b>Filtering results for FINERAD (non-thinned) dataset </b>: <br><br>
<b>SNPs filtered out</b> = 25,091 <br>
<b>SNPs retained </b> = 101,932 <br>

The next step of the filtering process entails omitting SNPs with heterozygosity greater than 0.45 to avoid potential additional bias due to paralogs (excess heterozygosity may be indicate of highly copied/duplicated loci in the genome). To identify snps with heterozygosity greater than 0.45, we use the software program GATK to generate SNP tables, which we use in R directly to calculate heterozygosity values and filter out SNPs.

### STEP 8: Generate a general bash shell script header for all GATK commands with which to run all GATK commands
In this step, we are generating a universal bash shell script to generate SNP tables with the program GATK. Here, we only generate the annotated header for the shell script, and in subsequent steps, we include and provide details on each line of code in the Rmd chunks. The shell script for running all GATK commands is called "gatk.sh".

#### 8a: Generate the header for a general bash shell script to run GATK:

##### Generate vcftools bash shell script: `code/shell_scripts/gatk.sh`
```{bash}
#! /bin/bash

#SBATCH -p Lewis  # use the Lewis partition
#SBATCH -J get_snp_table  # give the job a custom name
#SBATCH -o snp_table_vcf_allfilters-%j.out # give the job output a custom name
#SBATCH -t 0-02:00  # two hour time limit

#SBATCH -N 1  # number of nodes
#SBATCH -n 1  # number of cores (AKA tasks)

# Load GATK module

# module load rss/rss-2020
# module load gatk/gatk-4.0.1.1

# Commands with srun will run on all cores in the allocation
```

### STEP 9: Generate SNP tables for the popgen (thinned) and finerad (non-thinned) datasets.
In this step, we are generating SNP tables for the popgen (thinned) and finerad (non-thinned) datasets to compute allele and genotype frequencies for all SNPs across individuals. Specifically, we are computing heterozygosity at all SNPs to identify those with excess heterozygosity (greater than 0.45 across samples). 

#### 9a: Generate SNP table for POPGEN dataset 

##### 9a.1. Copy and paste the code below in the shell file generated in Step 8a above, which will generate a .txt file with allele counts for all SNPs:

##### Command line code for obtaining SNP allele and genotype frequency information. UNCOMMENT this code in the shell script:
```{bash}
#gatk VariantsToTable \
       # -V ../../data/processed_vcf/A_02_popgen_filter_mac.vcf \
      #  -F CHROM \
       # -F POS \
      #  -F ID \
       # -F QUAL \
      #  -F AC \
       # -F HET \
      #  -F HOM-REF \
       # -F HOM-VAR \
      #  -F NO-CALL \
       # -F VAR \
      #  -F NCALLED \
      #  -O ../../data/filtering_data/A_02_popgen_snps.txt
```

Run: `sbatch gatk.sh`

This code generates the file: `data/filtering_data/A_02_popgen_snps.txt`. This file should be used to compute genotype frequencies and generate a separate list of SNP IDs with excess heterozygosity.

#### 9a.2. Compute genotype frequencies and generate a list of SNP IDs for SNPs with excess heterozygosity for POPGEN dataset.

##### Generate list of sample IDs: `data/filtering_data/smb_genomics_popgen_het45.txt`
```{r}
## read in .txt file
popgen_snps <- read_tsv("data/filtering_data/A_02_popgen_snps.txt")

## Convert to data frame and mutate individual to factor
popgen_snps <-popgen_snps %>%
  mutate(snp_het = HET/NCALLED)

## Generate list of samples above missing genotype threshold
popgen_excess_het <- popgen_snps %>%
  filter(snp_het >= 0.45)

# Select the first column and convert to data frame
popgen_excess_het <- data.frame(popgen_excess_het[,c(1:2)])

# Save the list of bad samples as a .txt file without column names
write_tsv(popgen_excess_het, 
          file = "data/filtering_data/smb_genomics_popgen_het45.txt", 
          col_names = FALSE)
```

#### 9b: Generate SNP table for FINERAD dataset 

##### 9b.1. Copy and paste the code below in the shell file generated in Step 8a above, which will generate a .txt file with allele counts for all SNPs:

##### Command line code for obtaining SNP allele and genotype frequency information. UNCOMMENT this code in the shell script:
```{bash}
#gatk VariantsToTable \
    #    -V ../../data/processed_vcf/B_01_finerad_filter_mac.vcf \
    #    -F CHROM \
     #   -F POS \
    #    -F ID \
     #   -F QUAL \
    #    -F AC \
     #   -F HET \
    #    -F HOM-REF \
     #   -F HOM-VAR \
    #    -F NO-CALL \
     #   -F VAR \
    #    -F NCALLED \
     #   -O ../../data/filtering_data/B_01_finerad_snps.txt
```

Run: `sbatch gatk.sh`

This code generates the file: `data/filtering_data/B_01_finerad_snps.txt`. This file should be used to compute genotype frequencies and generate a separate list of SNP IDs with excess heterozygosity.

#### 9b.2. Compute genotype frequencies and generate a list of SNP IDs for SNPs with excess heterozygosity for FINERAD dataset.

##### Generate list of sample IDs: `data/filtering_data/smb_genomics_popgen_het45.txt`
```{r}
## read in .txt file
finerad_snps <- read_tsv("data/filtering_data/B_01_finerad_snps.txt")

## Convert to data frame and mutate individual to factor
finerad_snps <-finerad_snps %>%
  mutate(snp_het = HET/NCALLED)

## Generate list of samples above missing genotype threshold
finerad_excess_het <- finerad_snps %>%
  filter(snp_het >= 0.45)

# Select the first column and convert to data frame
finerad_excess_het <- data.frame(finerad_excess_het[,c(1:2)])

# Save the list of bad samples as a .txt file without column names
write_tsv(finerad_excess_het, 
          file = "data/filtering_data/smb_genomics_finerad_het45.txt", 
          col_names = FALSE)
```

### STEP 10: Filter out SNPs with greater than 45% heterozygosity across samples.
In this step, we are removing the SNPs with greater than 45% heterozygosity identified in Steps 9a.2 (for the thinned popgen dataset) and 9b.2 (for the non-thinned finerad dataset) above. 

#### 10a: Filter excess heterozygosity SNPs from POPGEN dataset

##### 10a.1. Copy and paste the code below in the shell file generated in Step 2a above, which will eliminate the SNPs with excess heterozygosity identified in Steps 9a.2 and 9b.2. above:

##### Command line code for omitting bad samples. UNCOMMENT this code in the shell script:
```{bash}
#vcftools --vcf ../../data/processed_vcf/A_02_popgen_filter_mac.vcf --exclude-positions ../../data/filtering_data/smb_genomics_popgen_het45.txt --recode --recode-INFO-all --out ../../data/processed_vcf/A_03_popgen_filter_het45
```

Run: `sbatch vcftools.sh`

This code generates the file: `data/processed_vcf/A_03_popgen_filter_het45.vcf`. This file should be used as input in the next filtering step.

<b>Filtering results for POPGEN (thinned) dataset </b>: <br><br>
<b>SNPs filtered out</b> = 934 <br>
<b>SNPs retained </b> = 50,828 <br>

#### 10b: Filter excess heterozygosity SNPs from FINERAD dataset

##### 10b.1. Copy and paste the code below in the shell file generated in Step 2a above, which will eliminate the SNPs with excess heterozygosity identified in Steps 9a.2 and 9b.2. above:

##### Command line code for omitting bad samples. UNCOMMENT this code in the shell script:
```{bash}
#vcftools --vcf ../../data/processed_vcf/B_01_finerad_filter_mac.vcf --exclude-positions ../../data/filtering_data/smb_genomics_finerad_het45.txt --recode --recode-INFO-all --out ../../data/processed_vcf/A_03_finerad_filter_het45
```

Run: `sbatch vcftools.sh`

This code generates the file: `data/processed_vcf/B_02_finerad_filter_het45.vcf`. This file should be used as input in the next filtering step.

<b>Filtering results for POPGEN (thinned) dataset </b>: <br><br>
<b>SNPs filtered out</b> = 3,273 <br>
<b>SNPs retained </b> = 98,659 <br>

At this point, the popgen (thinned) and finerad (non-thinned) datasets are fully filtered and ready for downstream population genomic analyses.

### STEP 11: Save the two fully filtered data files generated in Steps 10a.1 and 10b.1 above as new files ready to use for downstream population genomic analyses and save them here: `data/processed_vcf/`

#### 11a: Save POPGEN data; uncomment execute the following command in the shell to generate new file: `data/processed_vcf/popgen.vcf`
```{bash}
#cp A_03_popgen_filter_het45 ../analysis_data/popgen.vcf
```

#### 11a: Save FINERAD data; uncomment execute the following command in the shell to generate new file: `data/processed_vcf/finerad.vcf`
```{bash}
#cp B_02_finerad_filter_het45 ../analysis_data/finerad.vcf
```

### ----------------------- END OF PHASE 1: FILTER RAW VCF FILE ----------------------- ###
