---
title: 'Analysis 4: Population Inference'
author: "Joe Gunn"
date: "2022-08-29"
output: html_document
---

# Project: Population genomic analysis of Smallmouth Bass and Neosho Bass in the Central Interior Highlands
We investigated the extent of genomic divergence, local directional selection, and admixture between the Smallmouth Bass (<i>Micropterus dolomieu</i>) and the Neosho Bass (<i>M. velox</i>) in the Central Interior Highlands (CIH) ecoregion of central north America. Specifically, we used ddRADseq data to assessed the phylogenomic relationship between and within species, characterizing inter- and intraspecific diversity and SNPs potentially under local directional selection at the population level. Additionally, we inferred the relative timing of admxiture in Neosho Bass streams where there is known introgressive hybridization with Smallmouth Bass to understand the influence of natural, historic geographic factors on mixing (stream capture or transient flooding) vs. anthropogenic factors (i.e., non-native introductions through stocking), which is known to have occurred widely in these economically valuable species. We ultimately hoped to provide novel insights into the diversity of endemic, ecologically important and popular sport fish in the CIH.

## Specific Aim: Population Inference
In this analysis, we used the finerad.vcf data generated in Analysis 2 (SNP Filtering...) to assess fine-scale coancestry between Smallmouth Bass, and Neosho Bass in the CIH using haplotype inference (excluding Spotted Bass). Specifically, we estimated coancestry in 1) the full dataset, with all pure and admixed individuals, excluding the Spotted Bass X Smallmouth Bass hybrid (BFC10) inferred from population genomic analysis in Analysis 3, 2) the pure dataset, with only pure individuals of Smallmouth Bass and Neosho Bass, and 3) the admixed dataset, with only admixed individuals of Neosho Bass (no admixed Smallmouth Bass were detected).

## Phases of analysis:
### Phase 1: Data preparation 
### Phase 2: Coancestry analysis
### Phase 3: Plotting

## Libraries needed for analysis
```{r libraries, echo = FALSE}
library(readxl)
library(tidyverse)
library(cowplot)
library(devtools)
library(vcfR)
library(XML)
library(RColorBrewer)
source("code/source_code/FinestructureLibrary.R") #Fine-structure R source code
```

## PHASE 1: DATA PREPARATION
In this phase of the analysis, we are reading in the fully filtered, non-thinned VCF dataset (`finerad.vcf`) and further filtering these data to create three additional datasets, including 1) all Smallmouth Bass and Neosho Bass samples; 2) pure samples only, and 3) admixed samples only.

All analyses conducted by calling commands in bash shell scripts were performed on the Lewis high-performance computing cluster at the University of Missouri <br>

<b>Programs needed</b>: <br><br>

FINERADSTRUCTURE v.0.3.2 (Malinsky, Trucchi, Lawson, & Falush, 2018) <br>

<b>Citation</b>:<br>

Malinsky, M., Trucchi, E., Lawson, D. J., & Falush, D. (2018). RADpainter and fineRADstructure: Population inference from RADseq data. Molecular Biology and Evolution, 35, 1284â€“1290. doi:10.1093/molbev/msy023

### STEP 1: Filter out Spotted Bass samples and the Spotted Bass X Neosho Bass hybrid. 
In this step, we are filtering out Spotted Bass and the single Spotted Bass X Neosho Bass hybrid (BFC10) from the full, non-thinned dataset, since we are interested in assessing coance

#### 1a: Generate a general bash shell script header for data filtering in VCFTools
In this step, we are generating another bash shell script to run individual sample filtering in VCFTools (see `filtering_processing_analysis/smb_genomics_filtering_processing_analysis.Rmd` for details on programs needed, etc.). Here, we only generate the annotated header for the shell script, and in subsequent steps, we include and provide details on each data filtering step and the associated lines of code in Rmd chunks. The shell script for running all VCFTools code is called "vcftools.sh".

##### 1a.1: Copy and paste the code below in a new shell script file: 

##### Generate VCFTools bash shell script: `code/shell_scripts/vcftools.sh`
```{bash}
#! /bin/bash

#SBATCH -p Lewis  # use the Lewis partition
#SBATCH -J vcftools  # give the job a custom name
#SBATCH -o filter_spotted.out  # give the job output a custom name
#SBATCH -t 0-02:00  # two hour time limit

#SBATCH -N 1  # number of nodes
#SBATCH -n 1  # number of cores (AKA tasks)

# Load VCFtools module

# module load rss/rss-2020
# module load vcftools/vcftools-v0.1.14
```

##### Important Note: the generated file will have an additional extension (.recode), which is not necessary for downstream analyses. Every time a new VCF file was filtered, we manually removed this piece of the extension by copying the file to a new name (see below) and deleting the .recode.vcf file.

#### 1b: Filter the full, non-thinned VCF to remove Spotted Bass hybrid (BFC10) and Spotted Bass individuals.

##### 1b.1. Read in the .Rda file saved in `smb_genomics_admixture_analysis.Rmd`; run the Rmd chunk below:

##### Generate list of Spotted Bass hybrid: `data/filtering_data/bfc10.txt` 
```{r}
# Load in the .Rda file generated in the admixture analysis
load("../admixture_analysis/data/filtering_data/bcf10.Rda")

BFC10 <- as.data.frame(BFC10)

# Save the list of bad samples as a .txt file without column names
write_tsv(BFC10, 
          file = "data/filtering_data/bfc10.txt", 
          col_names = FALSE)

# Save the list of bad samples as a .Rda file without column names
save(BFC10, file = "data/filtering_data/bcf10.Rda")
```

#### 1b.2. Copy and paste the code below in the shell file generated in Step 1a.1 above, which will generate a filtered VCF file without BFC10:

##### Command line code for filtering. UNCOMMENT this code in the shell script:
```{bash}
## filter 01: omit Spotted Bass X Neosho Bass hybrid (BFC10) 
#vcftools --vcf ../../data/processed_vcf/finerad.vcf --remove ../../data/filtering_data/bfc10.txt --recode --recode-INFO-all --out ../../data/processed_vcf/01_finerad_spb_hybrid
```

Run: `sbatch vcftools.sh`

This code generates the file: `data/processed_bed/all_samples/01_finerad_spb_hybrid.vcf`. This file should be used as input in analysis in the next vcftools filter.

<b>Filtering results</b>: <br><br>
<b>Samples omitted </b> = 1 <br>
<b>Samples retained </b> = 91 <br>
<b>SNPs retained </b> = 50,828 <br>

##### 1b.3. Generate a list of only Spotted Bass to be used for filtering in subsequent steps:

##### Read in metadata and generate Spotted Bass list: `data/filtering_data/spotted_bass.txt`
```{r}
spotted_bass <- metadata %>%
  filter(species == "Spotted_Bass")

# Select the first column and convert to data frame
spotted_bass <- data.frame(spotted_bass) %>%
  select(sample_id)

# Save the list of bad samples as a .txt file without column names
write_tsv(spotted_bass, 
          file = "data/filtering_data/spotted_bass.txt", 
          col_names = FALSE)
```

##### 1b.2. Copy and paste the code below in the shell file generated in Step 1a.1 above, which will generate a filtered VCF file without Spotted Bass:

##### Command line code for filtering. UNCOMMENT this code in the shell script and run: `sbatch vcftools.sh`
```{bash}
## filter 02: omit Spotted Bass
#vcftools --vcf ../../data/processed_vcf/01_finerad_spb_hybrid.vcf --remove ../../data/filtering_data/spotted_bass.txt --recode --recode-INFO-all --out ../../data/processed_vcf/02_finerad_spb
```

Run: `sbatch vcftools.sh`

This code generates the file: `data/processed_bed/all_samples/02_finerad_spb.vcf`. This file should be used in the next vcftools filter.

<b>Filtering results</b>: <br><br>
<b>Samples omitted </b> = 4 <br>
<b>Samples retained </b> = 87 <br>
<b>SNPs retained </b> = 50,828 <br>

### STEP 2: Generate two datasets, one for only pure samples and one for only admixed samples
In this step, we are further filtering the finerad dataset to create datasets for 1) only pure samples and 2) only admixed samples. We are interested in understanding coancestry among individuals within these groups so that we may infer population designations based on haplotype similarity. Individuals that are closely related and likely within a single population form blocks of coancestry in a FINERADSTRUCTURE output heat map (see figures generated in PHASE 3 below).

#### 2b: Filter the full, non-thinned VCF to generate a dataset with all pure individuals.

##### 2b.1. Read in the .Rda file for admixed individuals saved in `smb_genomics_admixture_analysis.Rmd`; run the Rmd chunk below:

##### Generate list of admixed individuals (to be removed from the dataset): `data/filtering_data/admixed_individuals.txt` 
```{r}
# Load in the .Rda file generated in the admixture analysis
load("../admixture_analysis/data/filtering_data/admixed_individuals.Rda")

admixed_individuals <- as.data.frame(admixed_individuals)

# Save the list of bad samples as a .txt file without column names
write_tsv(admixed_individuals, 
          file = "data/filtering_data/admixed_individuals.txt", 
          col_names = FALSE)

# Save the list of bad samples as a .Rda file without column names
save(admixed_individuals, file = "data/filtering_data/admixed_individuals.Rda")
```

##### 2b.2. Copy and paste the code below in the shell file generated in Step 1a.1 above, which will generate a filtered VCF file with only pure individuals:

##### Command line code for filtering. UNCOMMENT this code in the shell script:
```{bash}
## filter A_03: Generate finerad dataset with only pure individuals
#vcftools --vcf ../../data/processed_vcf/02_finerad_spb.vcf --remove ../../data/filtering_data/admixed_individuals.txt --recode --recode-INFO-all --out ../../data/processed_vcf/A_03_finerad_pure
```

Run: `sbatch vcftools.sh`

This code generates the file: `data/processed_bed/all_samples/03_finerad_pure.vcf`. This file should be used for downstream analyses in FINERADSTRUCTURE.

<b>Filtering results</b>: <br><br>
<b>Samples omitted </b> = 40 <br>
<b>Samples retained </b> = 47 <br>
<b>SNPs retained </b> = 50,828 <br>

#### 2c: Filter the full, non-thinned VCF to generate a dataset with all admixed individuals.

##### 2c.1. Read in the .Rda file for pure individuals saved in `smb_genomics_admixture_analysis.Rmd`; run the Rmd chunk below:

##### Generate list of pure individuals (to be removed from the dataset): `data/filtering_data/pure_individuals.txt` 
```{r}
# Load in the .Rda file generated in the admixture analysis
load("../admixture_analysis/data/filtering_data/pure_individuals.Rda")

pure_individuals <- as.data.frame(pure_individuals)

# Save the list of bad samples as a .txt file without column names
write_tsv(pure_individuals, 
          file = "data/filtering_data/pure_individuals.txt", 
          col_names = FALSE)

# Save the list of bad samples as a .Rda file without column names
save(pure_individuals, file = "data/filtering_data/pure_individuals.Rda")
```

##### 2c.2. Copy and paste the code below in the shell file generated in Step 1a.1 above, which will generate a filtered VCF file with only pure individuals:

##### Command line code for filtering. UNCOMMENT this code in the shell script:
```{bash}
## filter B_03: Generate finerad dataset with only pure individuals
#vcftools --vcf ../../data/processed_vcf/02_finerad_spb.vcf --remove ../../data/filtering_data/pure_individuals.txt --recode --recode-INFO-all --out ../../data/processed_vcf/B_03_finerad_admixed
```

Run: `sbatch vcftools.sh`

This code generates the file: `data/processed_bed/all_samples/03_finerad_pure.vcf`. This file should be used for downstream analyses in FINERADSTRUCTURE.

<b>Filtering results</b>: <br><br>
<b>Samples omitted </b> = 47 <br>
<b>Samples retained </b> = 40 <br>
<b>SNPs retained </b> = 50,828 <br>

### ----------------------- END OF PHASE 1: DATA PREPARATION ----------------------- ###

## PHASE 2: COANCESTRY ANALYSIS
and generating haplotype mapping files (hapmap) files to be used for coancestry analysis with the software program FINERADSTRUCTURE

### STEP 1: Generate a general bash shell script header for all FINERADSTRUCTURE commands
In this step, we are generating a universal bash shell script to run all input file conversions and subsequent coancestry analyses in FINERADSTRUCTURE. Here, we only generate the annotated header for the shell script, and in subsequent steps, we include and provide details on each file conversion or analysis and the associated lines of code in Rmd chunks. The shell script for running all FINERADSTRUCTURE code is called "finerad.sh".

##### 1a: Generate the header for a general bash shell script to run FINERADSTRUCTURE:

##### Generate FINERADSTRUCTURE bash shell script: `code/shell_scripts/finerad.sh`
```{bash}
#! /bin/bash

#SBATCH -p Lewis  # use the Lewis partition
#SBATCH -J finerad  # give the job a custom name
#SBATCH -o finerad.out  # give the job output a custom name
#SBATCH -t 0-02:00  # two hour time limit

#SBATCH -N 1  # number of nodes
#SBATCH -n 1  # number of cores (AKA tasks)

# Commands here run only on the first core

# module load rss/rss-2020
# module load fineradstructure/fineradstructure-0.3.2
```

#### 1b: Generate painted hap map input files for coancestry analysis
In this step, we are converting VCF files to hap map files, which are input for the FINERADSTRUCTURE program, for 1) all Smallmouth Bass and Neosho Bass samples, 2) all pure samples only, and 3) all admixed samples.

##### 1b.1 Copy and paste the code below in the shell file generated in Step 1a above, which will generate hap map files for all three groups of interest (all individuals excluding Spotted Bass, pure individuals, and admixed individuals):

##### Generate hap map files. UNCOMMENT this code in the shell script:
```{bash}
## file conversion 01: convert VCF into haplotype chunks for all Smallmouth Bass and Neosho Bass
#RADpainter hapsFromVCF ../../data/processed_vcf/finerad.vcf > ../../datal/hap_maps/smb_neosho.txt
#RADpainter paint ../../data/hap_maps/smb_neosho.txt

## file conversion 02: convert VCF into haplotype chunks for all pure samples
#RADpainter hapsFromVCF ../../data/processed_vcf/A_03_finerad_pure.vcf > ../../datal/hap_maps/pure.txt
#RADpainter paint ../../data/hap_maps/pure.txt

## file conversion 03: convert VCF into haplotype chunks for all admixed samples
#RADpainter hapsFromVCF ../../data/processed_vcf/B_03_finerad_admixed.vcf > ../../datal/hap_maps/admixed.txt
#RADpainter paint ../../data/hap_maps/admixed.txt
```

#### 1c: Calculate coancestry matrices and build phylogenetic trees based on haplotype inference from hap map files

##### Copy and paste the code below in the shell file generated in Step 1a above, which will generate coancestry matrices and accompanying phylogenetic tree files for all three groups of interest (all individuals excluding Spotted Bass, pure individuals, and admixed individuals):

##### Calculate coancestry and generate phylogenies: 
```{bash}
## Coancestry 01: Smallmouth Bass and Neosho Bass
#finestructure -x 100000 -y 100000 -z 1000 ../../hap_maps/01_smb_neosho_chunks.out ../../mcmc_files/01_smb_neosho_chunks.mcmc.xml
#finestructure -m T -x 10000 ../../hap_maps/01_smb_neosho_chunks.out ../../mcmc_files/01_smb_neosho_chunks.mcmc.xml 01_smb_neosho_chunks.mcmcTree.xml

## Coancestry 02: Pure samples
#finestructure -x 100000 -y 100000 -z 1000 ../../hap_maps/02_pure_chunks.out ../../mcmc_files/02_pure_chunks.mcmc.xml
#finestructure -m T -x 10000 ../../hap_maps/02_pure_chunks.out ../../mcmc_files/02_pure_chunks.mcmc.xml 02_pure_chunks.mcmcTree.xml

## Coancestry 03:  Smallmouth Bass and Neosho Bass
#finestructure -x 100000 -y 100000 -z 1000 ../../hap_maps/03_admixed_chunks.out ../../mcmc_files/03_admixed_chunks.mcmc.xml
#finestructure -m T -x 10000 ../../hap_maps/03_admixed_chunks.out ../../mcmc_files/03_admixed_chunks.mcmc.xml 03_admixed_chunks.mcmcTree.xml
```

