---
title: 'Analysis 4: Population Inference'
author: "Joe Gunn"
date: "2022-08-29"
output: html_document
---

# Project: Population genomic analysis of Smallmouth Bass and Neosho Bass in the Central Interior Highlands
We investigated the extent of genomic divergence, local directional selection, and admixture between the Smallmouth Bass (<i>Micropterus dolomieu</i>) and the Neosho Bass (<i>M. velox</i>) in the Central Interior Highlands (CIH) ecoregion of central north America. Specifically, we used ddRADseq data to assessed the phylogenomic relationship between and within species, characterizing inter- and intraspecific diversity and SNPs potentially under local directional selection at the population level. Additionally, we inferred the relative timing of admxiture in Neosho Bass streams where there is known introgressive hybridization with Smallmouth Bass to understand the influence of natural, historic geographic factors on mixing (stream capture or transient flooding) vs. anthropogenic factors (i.e., non-native introductions through stocking), which is known to have occurred widely in these economically valuable species. We ultimately hoped to provide novel insights into the diversity of endemic, ecologically important and popular sport fish in the CIH.

## Specific Aim: Population Inference
In this analysis, we used the finerad.vcf data generated in Analysis 2 (SNP Filtering...) to assess fine-scale coancestry between Smallmouth Bass, and Neosho Bass in the CIH using haplotype inference (excluding Spotted Bass). Specifically, we estimated coancestry in 1) the full dataset, with all pure and admixed individuals, excluding the Spotted Bass X Smallmouth Bass hybrid (BFC10) inferred from population genomic analysis in Analysis 3, 2) the pure dataset, with only pure individuals of Smallmouth Bass and Neosho Bass, and 3) the admixed dataset, with only admixed individuals of Neosho Bass (no admixed Smallmouth Bass were detected).

## Phases of analysis:
### Phase 1: Data preparation 
### Phase 2: Coancestry analysis
### Phase 3: Plotting

## Libraries needed for analysis
```{r libraries, echo = FALSE}
library(readxl)
library(tidyverse)
library(cowplot)
library(devtools)
library(SNPRelate)
library(vcfR)
library(XML)
library(RColorBrewer)
source("../fineradstructure_analysis/FinestructureLibrary.R") #Fine-structure R source code
```

## PHASE 1: DATA PREPARATION
In this phase of the analysis, we are reading in the fully filtered, non-thinned VCF dataset (`finerad.vcf`) and generating haplotype mapping files (hapmap) files to be used for coancestry analysis with the software program FINERADSTRUCTURE. We are interested in delimiting populations of tight coancestry within the Smallmouth Bass and Neosho Bass, so we exclude Spotted Bass from the dataset before preceding with coancestry analyses for 1) all Smallmouth Bass and Neosho Bass samples; 2) pure samples only, and 3) admixed samples only.

All analyses conducted by calling commands in bash shell scripts were performed on the Lewis high-performance computing cluster at the University of Missouri <br>

<b>Programs needed</b>: <br><br>

FINERADSTRUCTURE v.0.3.2 (Malinsky, Trucchi, Lawson, & Falush, 2018) <br>

<b>Citation</b>:<br>

Malinsky, M., Trucchi, E., Lawson, D. J., & Falush, D. (2018). RADpainter and fineRADstructure: Population inference from RADseq data. Molecular Biology and Evolution, 35, 1284â€“1290. doi:10.1093/molbev/msy023

### STEP 1: Generate a general bash shell script header for all FINERADSTRUCTURE commands
In this step, we are generating a universal bash shell script to run all input file conversions and subsequent coancestry analyses in FINERADSTRUCTURE. Here, we only generate the annotated header for the shell script, and in subsequent steps, we include and provide details on each file conversion or analysis and the associated lines of code in Rmd chunks. The shell script for running all FINERADSTRUCTURE code is called "finerad.sh".

##### 1a: Generate the header for a general bash shell script to run FINERADSTRUCTURE:

##### Generate FINERADSTRUCTURE bash shell script: `code/shell_scripts/finerad.sh`
```{bash}

#! /bin/bash

#SBATCH -p Lewis  # use the Lewis partition
#SBATCH -J finerad  # give the job a custom name
#SBATCH -o finerad.out  # give the job output a custom name
#SBATCH -t 0-02:00  # two hour time limit

#SBATCH -N 1  # number of nodes
#SBATCH -n 1  # number of cores (AKA tasks)

# Commands here run only on the first core

# module load rss/rss-2020
# module load fineradstructure/fineradstructure-0.3.2
```

### STEP 2: Filter out Spotted Bass individuals and generate pure sample and admixed sample datasets for the full, non-thinned VCF.
In this step, we are filtering out Spotted Bass from the full, non-thinned dataset, since we are interested in assessing coancestry among Smallmouth Bass and Neosho Bass in this analysis. Additionally, we are again creating datasets for 1) only pure samples and 2) only admixed samples, but for the non-thinned dataset.

#### 2a: Generate a general bash shell script header for data filtering in VCFTools
In this step, we are generating another bash shell script to run individual sample filtering in VCFTools (see `filtering_processing_analysis/smb_genomics_filtering_processing_analysis.Rmd` for details on programs needed, etc.). Here, we only generate the annotated header for the shell script, and in subsequent steps, we include and provide details on each data filtering step and the associated lines of code in Rmd chunks. The shell script for running all VCFTools code is called "vcftools.sh".

#### 2a: Generate the header for a general bash shell script to run VCFTools:

##### Generate VCFTools bash shell script: `code/shell_scripts/vcftools.sh`
```{bash}
#! /bin/bash

#SBATCH -p Lewis  # use the Lewis partition
#SBATCH -J vcftools  # give the job a custom name
#SBATCH -o filter_spotted.out  # give the job output a custom name
#SBATCH -t 0-02:00  # two hour time limit

#SBATCH -N 1  # number of nodes
#SBATCH -n 1  # number of cores (AKA tasks)

# Load VCFtools module

# module load rss/rss-2020
# module load vcftools/vcftools-v0.1.14
```

##### Important Note: the generated file will have an additional extension (.recode), which is not necessary for downstream analyses. Every time a new VCF file was filtered, we manually removed this piece of the extension by copying the file to a new name (see below) and deleting the .recode.vcf file.

#### 3b: Filter the full, non-thinned VCF to remove Spotted Bass and Spotted Bass hybrid (BFC10).

##### 3b.1. Read in metadata and generate a list of only Spotted Bass to be used for filtering in subsequent steps:

##### Read in metadata and generate Spotted Bass list: `data/filtering_data/spotted_bass.txt`
```{r}
# load in pure metadata, which was created in the smb_genomics_admixture_analysis.Rmd analysis
load("../raw_data/metadata.Rda")

# Filter out Spotted Bass hybrid (BCF10)
metadata <- metadata %>%
  filter(sample_id != "BFC10_1")

spotted_bass <- metadata %>%
  filter(species == "Spotted_Bass")

# Select the first column and convert to data frame
spotted_bass <- data.frame(spotted_bass) %>%
  select(sample_id)

# Save the list of bad samples as a .txt file without column names
write_tsv(spotted_bass, 
          file = "data/filtering_data/spotted_bass.txt", 
          col_names = FALSE)
```

##### 3b.2. Copy and paste the code below in the shell file generated in Step 3a above, which will generate a filtered VCF file without Spotted Bass:

##### Command line code for filtering. UNCOMMENT this code in the shell script and run: `sbatch vcftools.sh`
```{bash}
## filter 01: omit Spotted Bass
#vcftools --vcf ../../data/processed_vcf/popgen.vcf --remove ../../data/filtering_data/bfc10.txt --recode --recode-INFO-all --out ../../data/processed_vcf/01_popgen_spb_hybrid
```

<b>Filtering results</b>: <br><br>
<b>Samples omitted </b> = 1 <br>
<b>Samples retained </b> = 91 <br>
<b>SNPs retained </b> = 50,828 <br>










### STEP 2: Generate painted hap map input files for coancestry analysis
In this step, we are converting VCF files for 1) 

#### 2a: Convert VCF data into .bed format for admixture analysis in the software program ADMIXURE.

##### 2a.1. Copy and paste the code below in the shell file generated in Step 1a above, which will generate a .bed, along with .bim, .fam, .log, and .nosex files:

<b>We use the following flags in the code</b>: <br>
    --allow-extra-chr: this flag is necessary due to the data being in the format of contigs rather than chromosomes
    --vcf-half-call m: this flag treats any half genotype calls (e.g., '0/.' or './1') as missing data
    --make-bed: this specifies output format of .bed

##### Command line code for converting VCF to BED. UNCOMMENT this code in the shell script and run: `sbatch plink.sh`
```{bash}
#plink --vcf ../../data/processed_vcf/popgen.vcf --allow-extra-chr --vcf-half-call m --make-bed --out ../../data/processed_bed/all_samples/01_all
```

Run: `sbatch plink.sh`

This code generates the file: `data/processed_bed/all_samples/01_all.bed`, along with accompanying file types. This file should be used as input in analysis in ADMIXTURE for the full dataset.
